// @kepler.gl/deckgl-layers/dist/deckgl-extensions/filter-arrow-layer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _core = require("@deck.gl/core");
var _constants = _interopRequireDefault(require("@luma.gl/constants"));
var _filterShaderModule = _interopRequireDefault(require("./filter-shader-module"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2["default"])(o), (0, _possibleConstructorReturn2["default"])(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2["default"])(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var VALUE_FILTERED = 1;
var defaultProps = {
  getFiltered: {
    type: 'accessor',
    value: VALUE_FILTERED
  }
};
/**
 * FilterArrowExtension - a deck.gl extension to filter arrow layer
 *
 * A simple extension to filter arrow layer based on the result of CPU filteredIndex,
 * so we can avoid filtering on the raw Arrow table and recreating geometry attributes.
 * Specifically, an attribute `filtered` is added to the layer to indicate whether the feature has been Filtered
 * the shader module is modified to discard the feature if filtered value is 0
 * the accessor getFiltered is used to get the value of `filtered` based on the value `filteredIndex` in Arrowlayer
 */
var FilterArrowExtension = exports["default"] = /*#__PURE__*/function (_LayerExtension) {
  function FilterArrowExtension() {
    (0, _classCallCheck2["default"])(this, FilterArrowExtension);
    return _callSuper(this, FilterArrowExtension, arguments);
  }
  (0, _inherits2["default"])(FilterArrowExtension, _LayerExtension);
  return (0, _createClass2["default"])(FilterArrowExtension, [{
    key: "getShaders",
    value: function getShaders() {
      return {
        modules: [_filterShaderModule["default"]],
        defines: {}
      };
    }
  }, {
    key: "initializeState",
    value: function initializeState() {
      var attributeManager = this.getAttributeManager();
      if (attributeManager) {
        attributeManager.add({
          filtered: {
            size: 1,
            type: _constants["default"].FLOAT,
            accessor: 'getFiltered',
            shaderAttributes: {
              filtered: {
                divisor: 0
              },
              instanceFiltered: {
                divisor: 1
              }
            }
          }
        });
      }
    }
  }]);
}(_core.LayerExtension);
(0, _defineProperty2["default"])(FilterArrowExtension, "defaultProps", defaultProps);
(0, _defineProperty2["default"])(FilterArrowExtension, "extensionName", 'FilterArrowExtension');
