// @kepler.gl/utils/dist/time.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TileTimeInterval = exports.TIME_INTERVALS_ORDERED = exports.TIMELINE_MODES = exports.SAMPLE_TIMELINE = exports.LayerToFilterTimeInterval = void 0;
exports.filterIntervalOptions = filterIntervalOptions;
exports.getFilterMappedValue = getFilterMappedValue;
exports.getInitialInterval = getInitialInterval;
exports.getIntervalByTicks = getIntervalByTicks;
exports.getTimelineFromFilter = exports.getTimelineFromAnimationConfig = void 0;
exports.intervalToFunction = intervalToFunction;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _d3Array = require("d3-array");
var _moment = _interopRequireDefault(require("moment"));
var _constants = require("@kepler.gl/constants");
var _commonUtils = require("@kepler.gl/common-utils");
var _aggregation = require("./aggregation");
var _plot = require("./plot");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var TIMELINE_MODES = exports.TIMELINE_MODES = {
  inner: 'inner',
  outer: 'outer'
};
var TileTimeInterval = exports.TileTimeInterval = {
  YEAR: 'Y',
  MONTH: 'M',
  DAY: 'D',
  HOUR: 'H',
  MINUTE: 'T'
};
var TIME_INTERVALS_ORDERED = exports.TIME_INTERVALS_ORDERED = [TileTimeInterval.MINUTE, TileTimeInterval.HOUR, TileTimeInterval.DAY, TileTimeInterval.MONTH, TileTimeInterval.YEAR];
var LayerToFilterTimeInterval = exports.LayerToFilterTimeInterval = (0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, TileTimeInterval.MINUTE, _constants.INTERVAL['1-minute']), TileTimeInterval.HOUR, _constants.INTERVAL['1-hour']), TileTimeInterval.DAY, _constants.INTERVAL['1-day']), TileTimeInterval.MONTH, _constants.INTERVAL['1-month']), TileTimeInterval.YEAR, _constants.INTERVAL['1-year']);
var SAMPLE_TIMELINE = exports.SAMPLE_TIMELINE = {
  // value: [15], // represent 15% of the all width
  value: [5, 15],
  // represent start at 5% and ends at 15%
  domain: [1, 100],
  // represent the total domain
  speed: 1,
  enableInteraction: false,
  // can use interact with this timeline
  isAnimating: false,
  step: null,
  // @todo: giuseppe coverType: 'inner' | 'outer'
  mode: TIMELINE_MODES.inner
  //  ....
};
var getTimelineFromAnimationConfig = exports.getTimelineFromAnimationConfig = function getTimelineFromAnimationConfig(animationConfig) {
  var currentTime = animationConfig.currentTime,
    domain = animationConfig.domain,
    speed = animationConfig.speed,
    isAnimating = animationConfig.isAnimating,
    timeSteps = animationConfig.timeSteps,
    defaultTimeFormat = animationConfig.defaultTimeFormat,
    timeFormat = animationConfig.timeFormat,
    timezone = animationConfig.timezone;
  return {
    // @ts-expect-error
    value: (0, _commonUtils.toArray)(currentTime),
    enableInteraction: true,
    domain: domain,
    speed: speed,
    isAnimating: isAnimating || false,
    timeSteps: timeSteps,
    defaultTimeFormat: defaultTimeFormat,
    timeFormat: timeFormat,
    timezone: timezone,
    timeBins: null,
    marks: null
  };
};

// check if the data inherent default time interval

// https://github.com/d3/d3-scale/blob/732ed4b1cd5c643700571d1089c7deb8472242a6/src/time.js#L69
// given number of ticks, calculate a reasonable interval
function getIntervalByTicks(ticks, start, stop) {
  if (ticks === null) ticks = 10;
  var tickIntervals = Object.values(_constants.TICK_INTERVALS);
  var interval;
  var step;
  // If a desired tick count is specified, pick a reasonable tick interval
  // based on the extent of the domain and a rough estimate of tick size.
  // Otherwise, assume interval is already a time interval and use it.
  if (typeof ticks === 'number') {
    var target = Math.abs(stop - start) / ticks;
    var i = (0, _d3Array.bisector)(function (d) {
      return d.duration;
    }).right(tickIntervals, target);
    if (i === tickIntervals.length) {
      step = (0, _d3Array.tickStep)(start / _constants.durationYear, stop / _constants.durationYear, ticks);
      interval = 'year';
    } else if (i) {
      var tickInterval = tickIntervals[target / tickIntervals[i - 1].duration < tickIntervals[i].duration / target ? i - 1 : i];
      // @ts-ignore TODO/ib
      step = tickInterval.step;
      // @ts-ignore TODO/ib
      interval = tickInterval.interval;
    } else {
      step = Math.max((0, _d3Array.tickStep)(start, stop, ticks), 1);
      interval = 'millisecond';
    }
  }
  return "".concat(step, "-").concat(interval);
}

// get a  number of unique samples
function getUniqueSamples(values, count) {
  var i = -1;
  var samples = [];
  var sampleMap = {};
  while (i++ < values.length && samples.length < count) {
    var v = values[i];
    if (v !== undefined && v !== null && !sampleMap[v]) {
      sampleMap[v] = true;
      samples.push(v);
    }
  }
  return Object.values(samples);
}

/**
 * Given an array of epoch timestamp. sort it, if number of element
 * share the same time interval exceed thresholf, and total steps smaller than 100, return it, else return null
 * @param values
 */
function detectInterval() {
  var values = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var domain = arguments.length > 1 ? arguments[1] : undefined;
  var maxSteps = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 10;
  var threshold = 0.7;
  var sorted = values.sort(_d3Array.ascending);

  // get first 100 unique sorted ts
  var samples = getUniqueSamples(sorted, 100);
  if (samples.length < 2) {
    return null;
  }

  // get all intervals
  var intervals = samples.reduce(function (accu, d, i) {
    if (i > 0) {
      var duration = _moment["default"].duration(_moment["default"].utc(samples[i]).diff(_moment["default"].utc(samples[i - 1])));
      var _getDurationUnit = getDurationUnit(duration),
        _getDurationUnit2 = (0, _slicedToArray2["default"])(_getDurationUnit, 2),
        dur = _getDurationUnit2[0],
        c = _getDurationUnit2[1];
      accu.push("".concat(c, "-").concat(dur));
    }
    return accu;
  }, []);

  // find the most occured interval
  var occur = (0, _aggregation.getFrequency)(intervals);
  var maxOccr = Object.keys(occur).reduce(function (prev, key) {
    return occur[prev] >= occur[key] ? prev : key;
  }, Object.keys(occur)[0]);

  // if occurance passed threshold
  var mostOccur = occur[maxOccr] / intervals.length;
  if (mostOccur >= threshold) {
    var _maxOccr$split = maxOccr.split('-'),
      _maxOccr$split2 = (0, _slicedToArray2["default"])(_maxOccr$split, 2),
      step = _maxOccr$split2[0],
      dur = _maxOccr$split2[1];
    var durationSecond = _constants.DURATIONS[dur] * parseInt(step); // eslint-disable-line radix
    var totalSteps = (domain[1] - domain[0]) / durationSecond;
    if (totalSteps < maxSteps) {
      // duration function is .days interval is day
      return maxOccr.substring(0, maxOccr.length - 1);
    }
  }
  return null;
}

/**
 * mappedValue is saved to dataset.fields.filterProps
 * @param dataset {KeplerTable}
 * @param filter
 */
function getFilterMappedValue(dataset, filter) {
  var dataId = dataset.id;
  var fieldName = filter.name[filter.dataId.indexOf(dataId)];
  var field = dataset.getColumnField(fieldName);
  if (!field) {
    // eslint-disable-next-line no-console, no-undef
    console.warn("field ".concat(fieldName, " does not exist on dataset"));
    return null;
  }
  var mappedValue = (field.filterProps || {}).mappedValue;
  if (!mappedValue) {
    // eslint-disable-next-line no-console, no-undef
    console.warn("mappedValue doesnt exist on filter field ".concat(filter.name));
    return null;
  }
  return mappedValue;
}
/**
 * Find the round unit of given durmostOccurtion: x years | months | days
 * @param duration
 */
function getDurationUnit(duration) {
  var durFuncs = Object.keys(_constants.DURATIONS);
  for (var i = 0; i < durFuncs.length; i++) {
    var c = duration[durFuncs[i]]();
    if (c > 0) {
      return [durFuncs[i], c];
    }
  }
  return ['milliseconds', 1];
}
function intervalToFunction(id) {
  var _id$split = id.split('-'),
    _id$split2 = (0, _slicedToArray2["default"])(_id$split, 2),
    stepStr = _id$split2[0],
    interval = _id$split2[1];
  var step = parseInt(stepStr); // eslint-disable-line radix
  if (!step) {
    // eslint-disable-next-line no-console, no-undef
    console.warn('Step is not an integer');
    return null;
  }
  if (!_constants.TIME_INTERVALS[interval]) {
    // eslint-disable-next-line no-console, no-undef
    console.warn("Undefined time interval ".concat(interval));
    return null;
  }
  return _constants.TIME_INTERVALS[interval].every(step);
}

/**
 * Get initial interval from filter and datasets
 * @param filter
 * @param datasets
 * @returns
 */
function getInitialInterval(filter, datasets) {
  var domain = filter.domain;
  var mergeMappedValue = filter.dataId.reduce(function (accu, dataId) {
    var mappedValue = getFilterMappedValue(datasets[dataId], filter);
    if (!mappedValue) {
      return accu;
    }
    for (var i = 0; i < mappedValue.length; i++) {
      accu.push(mappedValue[i]);
    }
    return accu;
  }, []);

  // check if data has predefined interval
  var interval = detectInterval(mergeMappedValue, domain);
  if (!interval) {
    // @ts-expect-error need better types for domain
    var _domain = (0, _slicedToArray2["default"])(domain, 2),
      t0 = _domain[0],
      t1 = _domain[1];
    interval = getIntervalByTicks(_constants.BINS_LARGE, t0, t1);
  }
  return interval;
}

// Filter interval options by time filter domain
// max number of interval is 1000
/**
 *
 * @param options
 * @param domain
 */
function filterIntervalOptions(options, domain) {
  var maxBins = 1000;
  var minBins = 2;
  var timeSpan = domain[1] - domain[0];
  return options.filter(function (op) {
    var id = op.id;
    if (!_constants.TICK_INTERVALS[id]) {
      return false;
    }
    var interval = _constants.TICK_INTERVALS[id];

    // rough count on bins
    var count = timeSpan / (interval.step * interval.duration);
    return count >= minBins && count <= maxBins;
  });
}

/**
 * Get timeline from filter
 * @param filter TimeRangeFilter filter
 * @returns Timeline
 */
var getTimelineFromFilter = exports.getTimelineFromFilter = function getTimelineFromFilter(filter) {
  var value = filter.value,
    domain = filter.domain,
    speed = filter.speed,
    isAnimating = filter.isAnimating,
    step = filter.step,
    timeSteps = filter.timeSteps,
    defaultTimeFormat = filter.defaultTimeFormat,
    timeFormat = filter.timeFormat,
    timezone = filter.timezone,
    timeBins = filter.timeBins,
    animationWindow = filter.animationWindow,
    plotType = filter.plotType;
  return {
    value: value,
    enableInteraction: true,
    domain: domain,
    speed: speed,
    isAnimating: isAnimating,
    step: step,
    timeSteps: timeSteps,
    defaultTimeFormat: defaultTimeFormat,
    timeFormat: timeFormat,
    timezone: timezone,
    timeBins: timeBins,
    animationWindow: animationWindow,
    marks: (0, _plot.getBinThresholds)(plotType === null || plotType === void 0 ? void 0 : plotType.interval, domain)
  };
};
