// @kepler.gl/common-utils/dist/h3-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCentroid = getCentroid;
exports.getHexFields = void 0;
exports.getVertices = getVertices;
Object.defineProperty(exports, "h3GetResolution", {
  enumerable: true,
  get: function get() {
    return _h3Js.h3GetResolution;
  }
});
Object.defineProperty(exports, "h3IsValid", {
  enumerable: true,
  get: function get() {
    return _h3Js.h3IsValid;
  }
});
exports.idToPolygonGeo = idToPolygonGeo;
exports.isHexField = void 0;
var _h3Js = require("h3-js");
var _constants = require("@kepler.gl/constants");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

// get vertices should return [lon, lat]
function getVertices(_ref) {
  var id = _ref.id;
  // always reverse it
  return (0, _h3Js.h3ToGeoBoundary)(id, true);
}

// get centroid should return [lon, lat]
function getCentroid(_ref2) {
  var id = _ref2.id;
  // always reverse it to [lng, lat]
  return (0, _h3Js.h3ToGeo)(id).reverse();
}
function idToPolygonGeo(object, properties) {
  if (!(object !== null && object !== void 0 && object.id)) {
    return null;
  }
  var vertices = getVertices(object);
  return {
    type: 'Feature',
    geometry: {
      coordinates: properties !== null && properties !== void 0 && properties.isClosed ? [vertices] : vertices,
      type: properties !== null && properties !== void 0 && properties.isClosed ? 'Polygon' : 'LineString'
    },
    properties: properties
  };
}
var isHexField = exports.isHexField = function isHexField(field, _fieldIdx, _dataContainer) {
  return field.type === _constants.ALL_FIELD_TYPES.h3;
};
var getHexFields = exports.getHexFields = function getHexFields(fields, dataContainer) {
  return fields.filter(function (f, i) {
    return isHexField(f, i, dataContainer);
  });
};
