// @kepler.gl/common-utils/dist/url.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isPMTilesUrl = void 0;
exports.parseUri = parseUri;
exports.validateUrl = validateUrl;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

/**
 * Allows to break down a url into multiple params
 * from http://blog.stevenlevithan.com/archives/parseuri
 */
function parseUri(str) {
  var o = parseUri.options;
  var m = o.parser[o.strictMode ? 'strict' : 'loose'].exec(str);
  var uri = {};
  var i = 14;
  while (i--) uri[o.key[i]] = (m === null || m === void 0 ? void 0 : m[i]) || '';
  uri[o.q.name] = {};
  uri[o.key[12]].replace(o.q.parser, function ($0, $1, $2) {
    if ($1) uri[o.q.name][$1] = $2;
  });
  return uri;
}
parseUri.options = {
  strictMode: false,
  key: ['source', 'protocol', 'authority', 'userInfo', 'user', 'password', 'host', 'port', 'relative', 'path', 'directory', 'file', 'query', 'anchor'],
  q: {
    name: 'queryKey',
    parser: /(?:^|&)([^&=]*)=?([^&]*)/g
  },
  parser: {
    strict:
    // eslint-disable-next-line no-useless-escape
    /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/,
    loose:
    // eslint-disable-next-line no-useless-escape
    /^(?:(?![^:@]+:[^:@\/]*@)([^:\/?#.]+):)?(?:\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/
  }
};

/**
 * Validates an url
 * @param str
 */
function validateUrl(str) {
  try {
    new URL(str);
    return true;
  } catch (_unused) {
    return false;
  }
}

/**
 * Checks whether a given URL points to a PMTiles file.
 * @param url The URL to check.
 * @returns True if the URL includes '.pmtiles', otherwise false.
 */
var isPMTilesUrl = exports.isPMTilesUrl = function isPMTilesUrl(url) {
  return url === null || url === void 0 ? void 0 : url.includes('.pmtiles');
};
