// @kepler.gl/deckgl-layers/dist/raster/webgl/pansharpen/pansharpen-brovey.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pansharpenBrovey = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

// Brovey Method: Each resampled, multispectral pixel is
// multiplied by the ratio of the corresponding
// panchromatic pixel intensity to the sum of all the
// multispectral intensities.
// Original code from https://github.com/mapbox/rio-pansharpen
//
var fs1 = "uniform sampler2D bitmapTexturePan;\nuniform float panWeight;\n\nfloat pansharpen_brovey_ratio(vec4 rgb, float pan, float weight) {\n  return pan / ((rgb.r + rgb.g + rgb.b * weight) / (2. + weight));\n}\n\nvec4 pansharpen_brovey_calc(vec4 rgb, float pan, float weight) {\n  float ratio = pansharpen_brovey_ratio(rgb, pan, weight);\n  return ratio * rgb;\n}\n";
var fs2 = "precision mediump usampler2D;\n\n#ifdef SAMPLER_TYPE\n  uniform SAMPLER_TYPE bitmapTexturePan;\n#else\n  uniform sampler2D bitmapTexturePan;\n#endif\n\nuniform float panWeight;\n\nfloat pansharpen_brovey_ratio(vec4 rgb, float pan, float weight) {\n  return pan / ((rgb.r + rgb.g + rgb.b * weight) / (2. + weight));\n}\n\nvec4 pansharpen_brovey_calc(vec4 rgb, float pan, float weight) {\n  float ratio = pansharpen_brovey_ratio(rgb, pan, weight);\n  return ratio * rgb;\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var imagePan = opts.imagePan,
    _opts$panWeight = opts.panWeight,
    panWeight = _opts$panWeight === void 0 ? 0.2 : _opts$panWeight;
  if (!imagePan) {
    return null;
  }
  return {
    bitmapTexturePan: imagePan,
    panWeight: panWeight
  };
}
var pansharpenBrovey = exports.pansharpenBrovey = {
  name: 'pansharpen_brovey',
  fs1: fs1,
  fs2: fs2,
  defines: {
    SAMPLER_TYPE: 'sampler2D'
  },
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    float pan_band = float(texture2D(bitmapTexturePan, coord).r);\n    image = pansharpen_brovey_calc(image, pan_band, panWeight);\n    "
  }
};
