// @kepler.gl/deckgl-layers/dist/raster/webgl/texture/combine-bands.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.combineBandsUint = exports.combineBandsInt = exports.combineBandsFloat = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var imageBands = opts.imageBands;
  if (!imageBands || imageBands.length === 0) {
    return null;
  }
  var _imageBands = (0, _slicedToArray2["default"])(imageBands, 4),
    bitmapTextureR = _imageBands[0],
    bitmapTextureG = _imageBands[1],
    bitmapTextureB = _imageBands[2],
    bitmapTextureA = _imageBands[3];

  // return default values to prevent deck.gl validation warnings.
  // Note: all bands must be set for the shadow effect to work as expected.
  return {
    bitmapTextureR: bitmapTextureR || bitmapTextureR,
    bitmapTextureG: bitmapTextureG || bitmapTextureR,
    bitmapTextureB: bitmapTextureB || bitmapTextureR,
    bitmapTextureA: bitmapTextureA || bitmapTextureR
  };
}
var fs1 = "uniform sampler2D bitmapTextureR;\nuniform sampler2D bitmapTextureG;\nuniform sampler2D bitmapTextureB;\nuniform sampler2D bitmapTextureA;\n";
var fs2 = "precision mediump float;\nprecision mediump int;\nprecision mediump usampler2D;\n\n#ifdef SAMPLER_TYPE\n  uniform SAMPLER_TYPE bitmapTextureR;\n  uniform SAMPLER_TYPE bitmapTextureG;\n  uniform SAMPLER_TYPE bitmapTextureB;\n  uniform SAMPLER_TYPE bitmapTextureA;\n#else\n  uniform sampler2D bitmapTextureR;\n  uniform sampler2D bitmapTextureG;\n  uniform sampler2D bitmapTextureB;\n  uniform sampler2D bitmapTextureA;\n#endif\n";
var combineBands = {
  name: 'combine-bands',
  fs1: fs1,
  fs2: fs2,
  getUniforms: getUniforms,
  defines: {
    SAMPLER_TYPE: 'sampler2D'
  },
  inject: {
    'fs:DECKGL_CREATE_COLOR': "\n    float channel1 = float(texture2D(bitmapTextureR, coord).r);\n    float channel2 = float(texture2D(bitmapTextureG, coord).r);\n    float channel3 = float(texture2D(bitmapTextureB, coord).r);\n    float channel4 = float(texture2D(bitmapTextureA, coord).r);\n\n    image = vec4(channel1, channel2, channel3, channel4);\n    "
  }
};
var combineBandsFloat = exports.combineBandsFloat = _objectSpread(_objectSpread({}, combineBands), {}, {
  name: 'combine-bands-float'
});
var combineBandsUint = exports.combineBandsUint = _objectSpread(_objectSpread({}, combineBands), {}, {
  name: 'combine-bands-uint',
  defines: {
    SAMPLER_TYPE: 'usampler2D'
  }
});
var combineBandsInt = exports.combineBandsInt = _objectSpread(_objectSpread({}, combineBands), {}, {
  name: 'combine-bands-int',
  defines: {
    SAMPLER_TYPE: 'isampler2D'
  }
});
