// @kepler.gl/deckgl-layers/dist/raster/webgl/texture/mask.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.maskUint = exports.maskInt = exports.maskFloat = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var inf = Math.pow(2, 62);
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var imageMask = opts.imageMask,
    maskKeepMin = opts.maskKeepMin,
    maskKeepMax = opts.maskKeepMax;
  if (!imageMask) {
    return null;
  }
  return {
    bitmapTextureMask: imageMask,
    uMaskKeepMin: Number.isFinite(maskKeepMin) ? maskKeepMin : -inf,
    uMaskKeepMax: Number.isFinite(maskKeepMax) ? maskKeepMax : inf
  };
}
var fs1 = "uniform sampler2D bitmapTextureMask;\nuniform float uMaskKeepMin;\nuniform float uMaskKeepMax;\n";
var fs2 = "precision mediump float;\nprecision mediump int;\nprecision mediump usampler2D;\n\n#ifdef SAMPLER_TYPE\n  uniform SAMPLER_TYPE bitmapTextureMask;\n#else\n  uniform sampler2D bitmapTextureMask;\n#endif\n\nuniform float uMaskKeepMin;\nuniform float uMaskKeepMax;\n";
var mask = {
  name: 'mask-image',
  fs1: fs1,
  fs2: fs2,
  getUniforms: getUniforms,
  defines: {
    SAMPLER_TYPE: 'sampler2D'
  },
  inject: {
    'fs:DECKGL_CREATE_COLOR': "\n    float mask_value = float(texture2D(bitmapTextureMask, coord).r);\n    if (mask_value < uMaskKeepMin) discard;\n    if (mask_value > uMaskKeepMax) discard;\n    "
  }
};
var maskFloat = exports.maskFloat = _objectSpread(_objectSpread({}, mask), {}, {
  name: 'mask-image-float'
});
var maskUint = exports.maskUint = _objectSpread(_objectSpread({}, mask), {}, {
  name: 'mask-image-uint',
  defines: {
    SAMPLER_TYPE: 'usampler2D'
  }
});
var maskInt = exports.maskInt = _objectSpread(_objectSpread({}, mask), {}, {
  name: 'mask-image-int',
  defines: {
    SAMPLER_TYPE: 'isampler2D'
  }
});
