// @kepler.gl/deckgl-layers/dist/raster/webgl/texture/reorder-bands.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.constructPermutationMatrix = constructPermutationMatrix;
exports.reorderBands = void 0;
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

/**
 * Reorder image bands on GPU
 * Uses a permutation matrix to reorder a vec4
 */

var fs = "uniform mat4 uReorder;\n\nvec4 reorder_image(vec4 image, mat4 ordering) {\n  return image.rgba * ordering;\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var ordering = opts.ordering;
  if (!ordering) {
    return null;
  }
  return {
    uReorder: constructPermutationMatrix(ordering)
  };
}

/**
 * Construct permutation matrix from vector
 *
 * @param vector  Vector describing how to reorder bands
 *
 * @return a mat4 permutation matrix representing how to reorder bands
 */
function constructPermutationMatrix(vector) {
  var mat4 = Array(16).fill(0);
  var row = 0;
  var _iterator = _createForOfIteratorHelper(vector),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var index = _step.value;
      mat4[row * 4 + index] = 1;
      row += 1;
    }

    // If input vector wasn't of length 4, add identity in final places
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  for (var r = row; r < 4; r++) {
    mat4[r * 4 + r] = 1;
  }
  return mat4;
}
var reorderBands = exports.reorderBands = {
  name: 'reorder-bands',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    image = reorder_image(image, uReorder);\n    "
  }
};
