// @kepler.gl/deckgl-layers/dist/raster/webgl/texture/rgba-image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rgbaImage = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var imageRgba = opts.imageRgba;
  if (!imageRgba) {
    return null;
  }
  return {
    bitmapTextureRgba: imageRgba
  };
}
var fs1 = "uniform sampler2D bitmapTextureRgba;\n";
var fs2 = "precision mediump float;\nprecision mediump int;\nprecision mediump usampler2D;\n\n#ifdef SAMPLER_TYPE\n  uniform SAMPLER_TYPE bitmapTextureRgba;\n#else\n  uniform sampler2D bitmapTextureRgba;\n#endif\n";
var rgbaImage = exports.rgbaImage = {
  name: 'rgba-image',
  fs1: fs1,
  fs2: fs2,
  getUniforms: getUniforms,
  defines: {
    SAMPLER_TYPE: 'sampler2D'
  },
  inject: {
    'fs:DECKGL_CREATE_COLOR': "\n    image = vec4(texture2D(bitmapTextureRgba, coord));\n    if (image.a < 0.5) {\n      discard;\n    }\n    "
  }
};
