// @kepler.gl/layers/dist/geojson-layer/geojson-info-modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _reactMarkdown = _interopRequireDefault(require("react-markdown"));
var _reactIntl = require("react-intl");
var _localization = require("@kepler.gl/localization");
var _exampleTable = _interopRequireDefault(require("../example-table"));
var _templateObject, _templateObject2, _templateObject3; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var InfoModal = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 13px;\n  color: ", ";\n\n  pre {\n    padding: 12px;\n    background-color: #f8f8f9;\n  }\n"])), function (props) {
  return props.theme.titleColorLT;
});
var StyledTitle = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 20px;\n  letter-spacing: 1.25px;\n  margin: 18px 0 14px 0;\n  color: ", ";\n"])), function (props) {
  return props.theme.titleColorLT;
});
var StyledCode = _styledComponents["default"].code(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n"])), function (props) {
  return props.theme.titleColorLT;
});
var exampleTableHeader = ['id', 'latitude', 'longitude', 'sort by'];
var exampleTabbleRows = [['A', '40.81773', '-74.20986', '0'], ['A', '40.81765', '-74.20987', '1'], ['A', '40.81746', '-74.20998', '2'], ['B', '40.64375', '-74.33242', '0'], ['B', '40.64353', '-74.20987', '1'], ['B', '40.64222', '-74.33001', '2']];
var ExampleTable = function ExampleTable() {
  return /*#__PURE__*/_react["default"].createElement(_exampleTable["default"], {
    className: "geojson-example-table"
  }, /*#__PURE__*/_react["default"].createElement("thead", null, /*#__PURE__*/_react["default"].createElement("tr", null, exampleTableHeader.map(function (v) {
    return /*#__PURE__*/_react["default"].createElement("th", {
      key: v
    }, v);
  }))), /*#__PURE__*/_react["default"].createElement("tbody", null, exampleTabbleRows.map(function (row, i) {
    return /*#__PURE__*/_react["default"].createElement("tr", {
      key: i
    }, row.map(function (v, j) {
      return /*#__PURE__*/_react["default"].createElement("td", {
        key: j
      }, /*#__PURE__*/_react["default"].createElement(StyledCode, null, v));
    }));
  })));
};
var GeojsonInfoModalFactory = function GeojsonInfoModalFactory(columnMode) {
  var GeojsonInfoModal = function GeojsonInfoModal() {
    var intl = (0, _reactIntl.useIntl)();
    return /*#__PURE__*/_react["default"].createElement(InfoModal, {
      className: "geojson-info-modal"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "geojson-info-modal__description"
    }, /*#__PURE__*/_react["default"].createElement(_reactMarkdown["default"], {
      children: intl.formatMessage({
        id: columnMode === 'geojson' ? 'modal.polygonInfo.description' : 'modal.polygonInfo.descriptionTable'
      })
    })), columnMode === 'table' ? /*#__PURE__*/_react["default"].createElement("div", {
      className: "geojson-info-modal__example"
    }, /*#__PURE__*/_react["default"].createElement(StyledTitle, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: "modal.polygonInfo.exampleTable"
    })), /*#__PURE__*/_react["default"].createElement(ExampleTable, null)) : null);
  };
  return GeojsonInfoModal;
};
var _default = exports["default"] = GeojsonInfoModalFactory;
