// @kepler.gl/schemas/dist/ui-state-schema.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.propertiesV1 = exports["default"] = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _versions = require("./versions");
var _schema = _interopRequireDefault(require("./schema"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var propertiesV1 = exports.propertiesV1 = {
  mapControls: new _schema["default"]({
    version: _versions.VERSIONS.v1,
    properties: {
      mapLegend: new _schema["default"]({
        version: _versions.VERSIONS.v1,
        properties: {
          active: null,
          settings: new _schema["default"]({
            version: _versions.VERSIONS.v1,
            properties: {
              position: null,
              contentHeight: null
            },
            key: 'settings'
          })
        },
        key: 'mapLegend'
      })
    },
    key: 'mapControls'
  })
};
var uiStateSchema = (0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, _versions.VERSIONS.v0, new _schema["default"]({
  version: _versions.VERSIONS.v0,
  properties: propertiesV1,
  key: 'uiState'
})), _versions.VERSIONS.v1, new _schema["default"]({
  version: _versions.VERSIONS.v1,
  properties: propertiesV1,
  key: 'uiState'
}));
var _default = exports["default"] = uiStateSchema;
