// @kepler.gl/constants/dist/color-palettes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PALETTE_TYPES = exports.KEPLER_COLOR_PALETTES = exports.DataVizColors = exports.D3_COLOR_CHROMATIC_SCHEME = exports.COLORBREWER_SCHEME = exports.CATEGORIES = void 0;
exports.buildCategoricalPalette = buildCategoricalPalette;
exports.colorPaletteToColorRange = colorPaletteToColorRange;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var d3ScaleChromatic = _interopRequireWildcard(require("d3-scale-chromatic"));
var _d3Array = require("d3-array");
var _chromaJs = _interopRequireDefault(require("chroma-js"));
var _console = _interopRequireDefault(require("global/console"));
var _d3Color = require("d3-color");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
// linear interpolator

var CATEGORIES = exports.CATEGORIES = {
  COLORBREWER: 'ColorBrewer',
  D3: 'D3',
  UBER: 'Uber',
  COLORBLIND: 'ColorBlind'
};
var PALETTE_TYPES = exports.PALETTE_TYPES = {
  SEQ: 'sequential',
  QUA: 'qualitative',
  DIV: 'diverging',
  CYC: 'cyclical'
};
var COLORBREWER_SCHEME = exports.COLORBREWER_SCHEME = (0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, PALETTE_TYPES.SEQ, ['BuGn', 'BuPu', 'GnBu', 'OrRd', 'PuBu', 'PuBuGn', 'PuRd', 'RdPu', 'YlGn', 'YlGnBu', 'YlOrBr', 'YlOrRd', 'Blues', 'Greens', 'Greys', 'Oranges', 'Purples', 'Reds']), PALETTE_TYPES.DIV, ['BrBG', 'PiYG', 'PRGn', 'PuOr', 'RdBu', 'RdGy', 'RdYlBu', 'RdYlGn', 'Spectral']), PALETTE_TYPES.QUA, ['Accent', 'Dark2', 'Paired', 'Pastel1', 'Pastel2', 'Set1', 'Set2', 'Set3']);

// https://rdrr.io/cran/RColorBrewer/man/ColorBrewer.html
var COLOR_BLIND_SAFE_MAP = {
  // colorbrewer
  BrBG: true,
  PiYG: true,
  PRGn: true,
  PuOr: true,
  RdBu: true,
  RdGy: false,
  RdYlBu: true,
  RdYlGn: false,
  Spectral: false,
  Accent: false,
  Dark2: true,
  Paired: true,
  Pastel1: false,
  Pastel2: false,
  Set1: false,
  Set2: true,
  Set3: false,
  Blues: true,
  BuGn: true,
  BuPu: true,
  GnBu: true,
  Greens: true,
  Greys: true,
  Oranges: true,
  OrRd: true,
  PuBu: true,
  PuBuGn: true,
  PuRd: true,
  Purples: true,
  RdPu: true,
  Reds: true,
  YlGn: true,
  YlGnBu: true,
  YlOrBr: true,
  YlOrRd: true,
  // d3 scale chromatic
  Sinebow: true,
  Rainbow: false,
  Turbo: true,
  Viridis: true,
  Inferno: true,
  Magma: true,
  Plasma: true,
  Cividis: true,
  Warm: true,
  Cool: false,
  CubehelixDefault: true,
  Tableau10: false
};
var D3_COLOR_CHROMATIC_SCHEME = exports.D3_COLOR_CHROMATIC_SCHEME = (0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, PALETTE_TYPES.CYC, ['Sinebow', 'Rainbow']), PALETTE_TYPES.SEQ, ['Turbo', 'Viridis', 'Inferno', 'Magma', 'Plasma', 'Cividis', 'Warm', 'Cool', 'CubehelixDefault']), PALETTE_TYPES.QUA, ['Tableau10']);
var DataVizColors = exports.DataVizColors = {
  aqua: '#12939A',
  tumbleweed: '#DDB27C',
  mule_fawn: '#88572C',
  tree_poppy: '#FF991F',
  flame: '#F15C17',
  sapphire: '#223F9A',
  orchid: '#DA70BF',
  chathams_blue: '#125C77',
  med_aquamarine: '#4DC19C',
  crocodile: '#776E57',
  java: '#17B8BE',
  chalky: '#F6D18A',
  light_taupe: '#B7885E',
  peach_orange: '#FFCB99',
  apricot: '#F89570',
  portage: '#829AE3',
  light_orchid: '#E79FD5',
  blue_green: '#1E96BE',
  bermuda: '#89DAC1',
  cloudy: '#B3AD9E'
};
var UberVizDiverging = {
  name: 'Uber Viz Diverging',
  category: CATEGORIES.UBER,
  type: PALETTE_TYPES.DIV,
  colors: ['#00939C', '#E6FAFA'],
  colors2: ['#FEEEE8', '#C22E00'],
  diverging: true,
  correctLightness: false,
  colorBlindSafe: true
};
var UberVizSequential = {
  name: 'Uber Viz Sequential',
  category: CATEGORIES.UBER,
  type: PALETTE_TYPES.SEQ,
  colors: ['#00939C', '#E6FAFA'],
  colorBlindSafe: true
};
var UberPool = {
  name: 'UberPool',
  type: PALETTE_TYPES.DIV,
  category: CATEGORIES.UBER,
  colors: ['#223F9A', '#CF1750', '#FAE300'],
  correctLightness: false,
  colorBlindSafe: true
};
var IceAndFire = {
  name: 'Ice And Fire',
  type: PALETTE_TYPES.DIV,
  category: CATEGORIES.UBER,
  colors: ['#0198BD', '#FAFEB3'],
  colors2: ['#FEEDB1', '#D50255'],
  diverging: true,
  colorBlindSafe: true
};
var GlobalWarming = {
  name: 'Global Warming',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#4C0035', '#AC1C17', '#FFC300'],
  colorBlindSafe: true
};
var Sunrise = {
  name: 'Sunrise',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#355C7D', '#C06C84', '#F8B195'],
  colorBlindSafe: true
};
var OceanGreen = {
  name: 'Ocean Green',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#37535E', '#3EACA8', '#E5EEC1'],
  colorBlindSafe: true
};
var PinkWine = {
  name: 'Pink Wine',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#2C1E3D', '#956485', '#EDD1CA'],
  colorBlindSafe: true
};
var PurpleBlueYellow = {
  name: 'Purple Blue Yellow',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#383C65', '#49838A', '#D6DEBF'],
  mode: 'hsl',
  colorBlindSafe: true
};
var ViovetOcean = {
  name: 'ViovetOcean',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#7400B8', '#5E60CE', '#4EA8DE', '#56CFE1', '#72EFDD'],
  colorBlindSafe: false
};
var SummerSky = {
  name: 'SummerSky',
  type: PALETTE_TYPES.SEQ,
  category: CATEGORIES.UBER,
  colors: ['#184E77', '#168AAD', '#76C893', '#D9ED92'],
  colorBlindSafe: false
};
var UberVizQualitative = {
  name: 'Uber Viz Qualitative',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.UBER,
  colors: Object.values(DataVizColors),
  colorBlindSafe: false
};

// https://personal.sron.nl/~pault/#sec:qualitative
// A set of Qualitative Colors designed by Paul Tol that are color blind friendly
var TolBright = {
  name: 'Tol Bright',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#4477AA', '#EE6677', '#228833', '#CCBB44', '#66CCEE', '#AA3377', '#BBBBBB'],
  colorBlindSafe: true
};
// Bad Data: #BBBBBB

var TolVibrant = {
  name: 'Tol Vibrant',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#EE7733', '#0077BB', '#33BBEE', '#EE3377', '#CC3311', '#009988', '#BBBBBB'],
  colorBlindSafe: true
};
// Bad Data: #BBBBBB

var TolMuted = {
  name: 'Tol Muted',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#CC6677', '#332288', '#DDCC77', '#117733', '#88CCEE', '#882255', '#44AA99', '#999933', '#AA4499'],
  colorBlindSafe: true
};
// Bad Data: #DDDDDD

var TolMediumContrast = {
  name: 'Tol Medium Contrast',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#6699CC', '#004488', '#EECC66', '#994455', '#997700', '#EE99AA'],
  colorBlindSafe: true
};
var TolLight = {
  name: 'Tol Light',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#77AADD', '#EE8866', '#EEDD88', '#FFAABB', '#99DDFF', '#44BB99', '#BBCC33', '#AAAA00'],
  colorBlindSafe: true
};

// https://jfly.uni-koeln.de/color/
var OkabeIto = {
  name: 'Okabe Ito',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#E69F00', '#56B4E9', '#009E73', '#F0E442', '#0072B2', '#D55E00', '#CC79A7', '#000000'],
  colorBlindSafe: true
};
// Bad Data: #DDDDDD

var FSQBrand = {
  name: 'FSQ Brand',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#3333FF', '#6166EB', '#2ED9C3', '#82E8DB', '#FCCC0A', '#FFDAAF', '#30A5D9', '#97DAF8'],
  colorBlindSafe: true
};
var FSQWarmTone = {
  name: 'FSQ Warm Tone',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#C00B05', '#D150A5', '#E98ECA', '#FECE5A', '#FFDDBF', '#FFB4D3', '#EE5D86', '#D8D2D2'],
  colorBlindSafe: true
};
var FSQCoolTone = {
  name: 'FSQ Cool Tone',
  type: PALETTE_TYPES.QUA,
  category: CATEGORIES.COLORBLIND,
  colors: ['#11439F', '#297EE8', '#95C6C9', '#FECE5A', '#FFDDBF', '#9FB1B7', '#5281B5', '#B9D0FB'],
  colorBlindSafe: true
};

/**
 * Build Categorical color palette
 */
function buildCategoricalPalette(_ref) {
  var name = _ref.name,
    category = _ref.category,
    colors = _ref.colors,
    colorBlindSafe = _ref.colorBlindSafe;
  var allColors;
  // find d3 color scheme
  var scheme = d3ScaleChromatic["scheme".concat(name)];
  if (!scheme && !colors) {
    _console["default"].warn("scheme".concat(name, " cant not be found in d3 scale chromatic, needs to provide colors"));
    allColors = ['#DDDDDD'];
  } else if (!scheme) {
    // build from colors
    allColors = colors;
  } else {
    allColors = scheme;
  }
  if (!allColors.length) {
    _console["default"].warn('Needs to provide valid d3 color scheme name or an array of colors');
  }
  return {
    name: name,
    category: category,
    type: PALETTE_TYPES.QUA,
    colorBlindSafe: colorBlindSafe !== null && colorBlindSafe !== void 0 ? colorBlindSafe : COLOR_BLIND_SAFE_MAP[name],
    maxStep: allColors.length,
    colors: function colors(numColors) {
      // if numColors > maxSteps,  will return allColors
      return allColors.slice(0, numColors).map(_colorToUppercase);
    }
  };
}
function _colorToUppercase(c) {
  return (0, _d3Color.color)(c).formatHex().toUpperCase();
}
/**
 * All sequantial palette is based on palette in d3-scale-chromatic
 * https://github.com/d3/d3-scale-chromatic/blob/main/src/index.js
 */
function buildSequentialPalette(_ref2) {
  var name = _ref2.name,
    type = _ref2.type,
    category = _ref2.category;
  if (!Object.prototype.hasOwnProperty.call(COLOR_BLIND_SAFE_MAP, name)) {
    _console["default"].warn("".concat(name, " does not exists in COLOR_BLIND_SAFE_MAP"));
  }
  var interpolator = d3ScaleChromatic["interpolate".concat(name)];
  return {
    name: name,
    type: type,
    category: category,
    colorBlindSafe: COLOR_BLIND_SAFE_MAP[name],
    colors: function colors(numColors) {
      return (0, _d3Array.range)(0, numColors, 1).map(function (d) {
        return interpolator(d / (numColors - 1));
      }).map(_colorToUppercase);
    },
    linear: function linear() {
      return interpolator;
    }
  };
}
function buildCustomPalette(_ref3) {
  var colors = _ref3.colors,
    _ref3$colors = _ref3.colors2,
    colors2 = _ref3$colors === void 0 ? [] : _ref3$colors,
    _ref3$correctLightnes = _ref3.correctLightness,
    correctLightness = _ref3$correctLightnes === void 0 ? true : _ref3$correctLightnes,
    _ref3$bezier = _ref3.bezier,
    bezier = _ref3$bezier === void 0 ? false : _ref3$bezier,
    _ref3$diverging = _ref3.diverging,
    diverging = _ref3$diverging === void 0 ? false : _ref3$diverging,
    _ref3$mode = _ref3.mode,
    mode = _ref3$mode === void 0 ? 'lch' : _ref3$mode,
    name = _ref3.name,
    type = _ref3.type,
    category = _ref3.category,
    colorBlindSafe = _ref3.colorBlindSafe;
  var palette = {
    name: name,
    type: type,
    category: category,
    colorBlindSafe: colorBlindSafe,
    colors: function colors() {
      return [];
    },
    linear: function linear() {
      return function () {
        return '';
      };
    }
  };
  if (!colors.length) {
    _console["default"].error('colors has to be an array of colors');
    return;
  }
  var scaleLeft = _chromaJs["default"].scale(bezier && colors.length > 1 ? _chromaJs["default"].bezier(colors) : colors).mode(mode).correctLightness(correctLightness);
  var scaleRight;
  var scaleFull;
  if (diverging) {
    if (!colors.length) {
      _console["default"].error('colors2 has to be an array of colors when diverging = true');
      return;
    }
    scaleRight = _chromaJs["default"].scale(bezier && colors2.length > 1 ? _chromaJs["default"].bezier(colors2) : colors2).mode(mode).correctLightness(correctLightness);
    scaleFull = _chromaJs["default"].scale(bezier ? _chromaJs["default"].bezier(colors.concat(colors2)) : colors.concat(colors2)).mode(mode).correctLightness(correctLightness);
  }

  // return numColors => (scaleLeft ? stepsLeft.colors(numColors) : []);

  // given number of colors return color steps
  palette.colors = function (numColors) {
    if (diverging) {
      var even = numColors % 2 === 0;
      var numColorsLeft = Math.ceil(numColors / 2) + (even ? 1 : 0);
      var numColorsRight = Math.ceil(numColors / 2) + (even ? 1 : 0);
      var colorsLeft = scaleLeft ? scaleLeft.colors(numColorsLeft) : [];
      var colorsRight = scaleRight ? scaleRight.colors(numColorsRight) : [];
      var _steps = (even ? colorsLeft.slice(0, colorsLeft.length - 1) : colorsLeft).concat(colorsRight.slice(1)).map(_colorToUppercase);
      return _steps;
    }
    return scaleLeft ? scaleLeft.colors(numColors).map(_colorToUppercase) : [];
  };
  palette.linear = function () {
    return diverging ? scaleFull : scaleLeft;
  };
  return palette;
}
function buildPaletteBySchemeGroups(schemeGroups, category) {
  return Object.entries(schemeGroups).reduce(function (accu, _ref4) {
    var _ref5 = (0, _slicedToArray2["default"])(_ref4, 2),
      type = _ref5[0],
      palettes = _ref5[1];
    return [].concat((0, _toConsumableArray2["default"])(accu), (0, _toConsumableArray2["default"])(palettes.reduce(function (group, name) {
      var colorPalette = type === PALETTE_TYPES.QUA ? buildCategoricalPalette({
        name: name,
        category: category
      }) :
      // @ts-ignore jsdoc cant figure out type of 'type'
      buildSequentialPalette({
        name: name,
        type: type,
        category: category
      });
      // @ts-ignore type is not assignable to never[]
      group.push(colorPalette);
      return group;
    }, [])));
  }, []);
}
var COLORBREWER_PALETTES = buildPaletteBySchemeGroups(COLORBREWER_SCHEME, CATEGORIES.COLORBREWER);
var D3_COLOR_PALETTES = buildPaletteBySchemeGroups(D3_COLOR_CHROMATIC_SCHEME, CATEGORIES.D3);
var BRANDED_PALETTES = [UberVizDiverging, UberVizSequential, UberPool, IceAndFire, GlobalWarming, Sunrise, OceanGreen, PinkWine, PurpleBlueYellow, ViovetOcean, SummerSky, UberVizQualitative, TolBright, TolVibrant, TolMuted, TolMediumContrast, TolLight, OkabeIto, FSQBrand, FSQWarmTone, FSQCoolTone].map(function (recipe) {
  return recipe.type === PALETTE_TYPES.QUA ? buildCategoricalPalette(recipe) : buildCustomPalette(recipe);
}).filter(Boolean);
var KEPLER_COLOR_PALETTES = exports.KEPLER_COLOR_PALETTES = [].concat((0, _toConsumableArray2["default"])(BRANDED_PALETTES), (0, _toConsumableArray2["default"])(COLORBREWER_PALETTES), (0, _toConsumableArray2["default"])(D3_COLOR_PALETTES));

/**
 * create color range from palette, with steps and reversed as config
 */
function colorPaletteToColorRange(colorPalette, colorConfig) {
  var steps = colorConfig.steps,
    reversed = colorConfig.reversed;
  var colors = colorPalette.colors(steps).slice();
  if (reversed) {
    colors.reverse();
  }
  return _objectSpread({
    name: colorPalette.name,
    type: colorPalette.type,
    category: colorPalette.category,
    colors: colors
  }, reversed ? {
    reversed: reversed
  } : {});
}
