// @kepler.gl/components/dist/side-panel/common/dataset-tag.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DatasetTagFactory;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _localization = require("@kepler.gl/localization");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _ = require("../..");
var _templateObject, _templateObject2; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function nop() {
  return;
}
var DatasetTagWrapper = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  color: ", ";\n  font-size: 11px;\n  letter-spacing: 0.2px;\n  overflow: auto;\n\n  .dataset-color {\n    flex-shrink: 0;\n    margin-top: 3px;\n  }\n\n  .dataset-name {\n    white-space: nowrap;\n    overflow: hidden;\n    text-overflow: ellipsis;\n  }\n"])), function (props) {
  return props.theme.textColor;
});
var DatasetColorPicker = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n"])));
var UpdateTableColor = function UpdateTableColor(_ref) {
  var children = _ref.children,
    id = _ref.id;
  return /*#__PURE__*/_react["default"].createElement(DatasetColorPicker, {
    className: "dataset-action update-color",
    "data-tip": true,
    "data-for": "update-color-".concat(id)
  }, children);
};
function DatasetTagFactory() {
  var DatasetTag = function DatasetTag(_ref2) {
    var _ref2$onClick = _ref2.onClick,
      onClick = _ref2$onClick === void 0 ? nop : _ref2$onClick,
      _ref2$onClickSquare = _ref2.onClickSquare,
      onClickSquare = _ref2$onClickSquare === void 0 ? nop : _ref2$onClickSquare,
      dataset = _ref2.dataset,
      updateTableColor = _ref2.updateTableColor;
    return /*#__PURE__*/_react["default"].createElement(DatasetTagWrapper, {
      className: "source-data-tag"
    }, /*#__PURE__*/_react["default"].createElement(UpdateTableColor, {
      id: dataset.id
    }, /*#__PURE__*/_react["default"].createElement(_.DatasetSquare, {
      className: "dataset-color",
      backgroundColor: dataset.color,
      onClick: onClickSquare,
      "data-tip": true,
      "data-for": "update-color-".concat(dataset.id)
    }), updateTableColor ? /*#__PURE__*/_react["default"].createElement(_.Tooltip, {
      id: "update-color-".concat(dataset.id),
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement("span", null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: 'Update color'
    }))) : null), /*#__PURE__*/_react["default"].createElement("div", {
      className: "dataset-name",
      title: dataset.label,
      onClick: onClick
    }, dataset.label));
  };
  return DatasetTag;
}
