// @kepler.gl/constants/dist/user-guides.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.USER_GUIDE_DOC = exports.MAPBOX_ACCESS_TOKEN = exports.GUIDES_FILE_FORMAT_DOC = exports.EXPORT_HTML_MAP_MODES_DOC = exports.EXPORT_HTML_MAP_DOC = exports.BUG_REPORT_LINK = exports.API_REFERENCE_DOC = exports.ADD_DATA_TO_MAP_DOC = exports.ACTIONS_DOC = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var USER_GUIDE_DOC = exports.USER_GUIDE_DOC = 'https://docs.kepler.gl/docs/user-guides';
var ACTIONS_DOC = exports.ACTIONS_DOC = 'https://docs.kepler.gl/docs/api-reference/actions/actions';
var API_REFERENCE_DOC = exports.API_REFERENCE_DOC = 'https://docs.kepler.gl/docs/api-reference';
var EXPORT_HTML_MAP_DOC = exports.EXPORT_HTML_MAP_DOC = "".concat(USER_GUIDE_DOC, "/k-save-and-export");
var ADD_DATA_TO_MAP_DOC = exports.ADD_DATA_TO_MAP_DOC = "".concat(ACTIONS_DOC, "#adddatatomap");
var MAPBOX_ACCESS_TOKEN = exports.MAPBOX_ACCESS_TOKEN = 'https://docs.mapbox.com/help/how-mapbox-works/access-tokens/';
var EXPORT_HTML_MAP_MODES_DOC = exports.EXPORT_HTML_MAP_MODES_DOC = "".concat(USER_GUIDE_DOC, "/k-save-and-export#export-html-map");
var GUIDES_FILE_FORMAT_DOC = exports.GUIDES_FILE_FORMAT_DOC = "".concat(USER_GUIDE_DOC, "/b-kepler-gl-workflow/a-add-data-to-the-map#supported-file-formats");
var BUG_REPORT_LINK = exports.BUG_REPORT_LINK = 'https://github.com/keplergl/kepler.gl/issues/new?assignees=&labels=bug&template=bug_report.md&title=%5BBug%5D';
