// @kepler.gl/utils/dist/data-row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DataRow = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
/**
 * Setting for shared row optimization.
 * - False/undefined indicates that unique row objects should be used (default).
 * - True indicates that a single temporary row object should be created and used without extra allocations.
 * - A DataRow object indicates that the row should be used as a temporary shared row.
 * When used, the content of the shared row isn't preserved between calls.
 */
/**
 * Return type for createSharedRow:
 * - DataRow object that should be used as shared row.
 * - Falsy values indicate that shared row object shouldn't be used.
 */
var DataRow = exports.DataRow = /*#__PURE__*/function () {
  /**
   * Creates new DataRow.
   * @param dataContainer Data container where data is stored. Can be initialized with null for shared rows.
   * @param rowIndex Index of a row in the data container.
   */
  function DataRow(dataContainer, rowIndex) {
    (0, _classCallCheck2["default"])(this, DataRow);
    (0, _defineProperty2["default"])(this, "_dataContainer", void 0);
    (0, _defineProperty2["default"])(this, "_rowIndex", void 0);
    this._dataContainer = dataContainer;
    this._rowIndex = rowIndex;
  }

  /**
   * Conditionally creates a DataRow object.
   * @param sharedRowDesc Accepts forllowing options:
   * - true indicates that new DataRow should be created.
   * - falsy value or a DataRow object is passed through without any change.
   * @returns A new DataRow object or unchanged input argument.
   */
  return (0, _createClass2["default"])(DataRow, [{
    key: "valueAt",
    value:
    /**
     * Returns the value stored at the specified index in the row.
     * @param columnIndex Index of the requested field in the row.
     * @returns Value at the index.
     */
    function valueAt(columnIndex) {
      var _this$_dataContainer;
      return (_this$_dataContainer = this._dataContainer) === null || _this$_dataContainer === void 0 ? void 0 : _this$_dataContainer.valueAt(this._rowIndex, columnIndex);
    }

    /**
     * Returns the row represented as an array.
     * @returns The row represented as an array.
     */
  }, {
    key: "values",
    value: function values() {
      return this._dataContainer ? this._dataContainer.rowAsArray(this._rowIndex) : [];
    }

    /**
     * Setup a row object. The method is used by shared rows to prevent excessive allocations.
     * @param dataContainer Data container.
     * @param rowIndex Index of a row in the data container.
     */
  }, {
    key: "setSource",
    value: function setSource(dataContainer, rowIndex) {
      this._dataContainer = dataContainer;
      this._rowIndex = rowIndex;
    }

    /**
     * Creates a new array populated with the results of calling the provided function
     * on every element of the row.
     * @param handler The callback is called with the following arguments:
     * - elem: The current element being processed in the row.
     * - index: The index of the current element being processed in the row.
     * @returns A new array with each element being the result of the func callback.
     */
  }, {
    key: "map",
    value: function map(handler) {
      var _this$_dataContainer2;
      var numColumns = ((_this$_dataContainer2 = this._dataContainer) === null || _this$_dataContainer2 === void 0 ? void 0 : _this$_dataContainer2.numColumns()) || 0;
      var out = [];
      for (var column = 0; column < numColumns; ++column) {
        out[column] = handler(this.valueAt(column), column);
      }
      return out;
    }
  }], [{
    key: "createSharedRow",
    value: function createSharedRow(sharedRowDesc) {
      if (sharedRowDesc === true) {
        return new DataRow(null, 0);
      }
      return sharedRowDesc;
    }
  }]);
}();
