// @kepler.gl/components/dist/layer-animation-controller.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireWildcard(require("react"));
var _constants = require("@kepler.gl/constants");
var _utils = require("@kepler.gl/utils");
var _commonUtils = require("@kepler.gl/common-utils");
var _animationController = _interopRequireDefault(require("./common/animation-control/animation-controller"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

LayerAnimationControllerFactory.deps = [_animationController["default"]];
function LayerAnimationControllerFactory(AnimationController) {
  var LayerAnimationController = function LayerAnimationController(_ref) {
    var animationConfig = _ref.animationConfig,
      setLayerAnimationTime = _ref.setLayerAnimationTime,
      children = _ref.children;
    var timeSteps = animationConfig.timeSteps,
      domain = animationConfig.domain;
    var setTimelineValue = (0, _react.useCallback)(function (value) {
      var timelineValue = (0, _commonUtils.toArray)(value)[0];
      if (Array.isArray(timeSteps)) {
        setLayerAnimationTime((0, _utils.snapToMarks)(timelineValue, timeSteps));

        // TODO: merge slider in to avoid this step
      } else if (domain && timelineValue >= domain[0] && timelineValue <= domain[1]) {
        setLayerAnimationTime(timelineValue);
      }
    }, [domain, setLayerAnimationTime, timeSteps]);
    var timeline = (0, _utils.getTimelineFromAnimationConfig)(animationConfig);
    return /*#__PURE__*/_react["default"].createElement(AnimationController, {
      key: "layer-control",
      value: Number(animationConfig.currentTime),
      domain: animationConfig.domain,
      speed: animationConfig.speed,
      isAnimating: animationConfig.isAnimating,
      steps: animationConfig.timeSteps,
      animationWindow: animationConfig.timeSteps ? _constants.ANIMATION_WINDOW.interval : _constants.ANIMATION_WINDOW.point,
      setTimelineValue: setTimelineValue,
      timeline: timeline,
      children: children
    });
  };
  return LayerAnimationController;
}
var _default = exports["default"] = LayerAnimationControllerFactory;
