// @kepler.gl/deckgl-layers/dist/raster/webgl/color/colormap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.colormap = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var fs = "uniform sampler2D uColormapTexture;\nuniform int uHasCategoricalColors;\nuniform int uCategoricalMinValue;\nuniform int uCategoricalMaxValue;\nuniform int uMaxPixelValue;\n\n// Apply colormap texture given value\n// Since the texture only varies in the x direction, setting v to 0.5 as a\n// constant is fine\n// Assumes the input range of value is -1 to 1\nvec4 colormap(sampler2D cmap, vec4 image) {\n  vec2 uv;\n  if (uHasCategoricalColors == 1) {\n    float step = float(uMaxPixelValue) / float(uCategoricalMaxValue - uCategoricalMinValue);\n    uv = vec2(image.r * step, 0.5);\n  } else {\n    uv = vec2(0.5 * image.r + 0.5, 0.5);\n  }\n  vec4 color = texture2D(cmap, uv);\n  if(color.a <= 0.0) discard;\n  return color;\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var imageColormap = opts.imageColormap,
    minCategoricalBandValue = opts.minCategoricalBandValue,
    maxCategoricalBandValue = opts.maxCategoricalBandValue,
    dataTypeMaxValue = opts.dataTypeMaxValue,
    maxPixelValue = opts.maxPixelValue;
  if (!imageColormap) {
    return null;
  }
  var isSupportedDataType = Number.isFinite(dataTypeMaxValue);
  var isCategorical = isSupportedDataType && Number.isFinite(maxPixelValue) && Number.isFinite(minCategoricalBandValue) && Number.isFinite(maxCategoricalBandValue);
  return {
    uColormapTexture: imageColormap,
    uHasCategoricalColors: isCategorical ? 1 : 0,
    uCategoricalMinValue: Number.isFinite(minCategoricalBandValue) ? minCategoricalBandValue : 0,
    uCategoricalMaxValue: Number.isFinite(maxCategoricalBandValue) ? maxCategoricalBandValue : 0,
    uMaxPixelValue: Number.isFinite(maxPixelValue) ? maxPixelValue : 0
  };
}
var colormap = exports.colormap = {
  name: 'colormap',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    image = colormap(uColormapTexture, image);\n    "
  }
};
