// @kepler.gl/deckgl-layers/dist/raster/webgl/color/filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filter = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var fs = "uniform float filterMin1;\nuniform float filterMax1;\nuniform float filterMin2;\nuniform float filterMax2;\nuniform float filterMin3;\nuniform float filterMax3;\nuniform float filterMin4;\nuniform float filterMax4;\n";

// You can't pass JS' -Infinity or Infinity to a shader as a uniform
var inf = Math.pow(2, 62);

// eslint-disable-next-line complexity
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var filterMin1 = opts.filterMin1,
    filterMin2 = opts.filterMin2,
    filterMin3 = opts.filterMin3,
    filterMin4 = opts.filterMin4,
    filterMax1 = opts.filterMax1,
    filterMax2 = opts.filterMax2,
    filterMax3 = opts.filterMax3,
    filterMax4 = opts.filterMax4;
  if (Number.isFinite(filterMin1) || Number.isFinite(filterMin2) || Number.isFinite(filterMin3) || Number.isFinite(filterMin4) || Number.isFinite(filterMax1) || Number.isFinite(filterMax2) || Number.isFinite(filterMax3) || Number.isFinite(filterMax4)) {
    return {
      filterMin1: Number.isFinite(filterMin1) ? filterMin1 : -inf,
      filterMin2: Number.isFinite(filterMin2) ? filterMin2 : -inf,
      filterMin3: Number.isFinite(filterMin3) ? filterMin3 : -inf,
      filterMin4: Number.isFinite(filterMin4) ? filterMin4 : -inf,
      filterMax1: Number.isFinite(filterMax1) ? filterMax1 : inf,
      filterMax2: Number.isFinite(filterMax2) ? filterMax2 : inf,
      filterMax3: Number.isFinite(filterMax3) ? filterMax3 : inf,
      filterMax4: Number.isFinite(filterMax4) ? filterMax4 : inf
    };
  }
  return null;
}
var filter = exports.filter = {
  name: 'filter',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    if (image.r < filterMin1) discard;\n    if (image.g < filterMin2) discard;\n    if (image.b < filterMin3) discard;\n    if (image.a < filterMin4) discard;\n    if (image.r > filterMax1) discard;\n    if (image.g > filterMax2) discard;\n    if (image.b > filterMax3) discard;\n    if (image.a > filterMax4) discard;\n    "
  }
};
