// @kepler.gl/deckgl-layers/dist/raster/webgl/color/gamma-contrast.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.gammaContrast = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

// Gamma correction is a nonlinear operation that
// adjusts the image's channel values pixel-by-pixel according
// to a power-law:
//
// .. math:: pixel_{out} = pixel_{in} ^ {gamma}
//
// Setting gamma (:math:gamma) to be less than 1.0 darkens the image and
// setting gamma to be greater than 1.0 lightens it.

// Parameters
// ----------
// gamma (:math:gamma): float
//     Reasonable values range from 0.8 to 2.4.

// NOTE: Input array must have float values between 0 and 1!
// NOTE: gamma must be >= 0
var fs = "#define epsilon 0.00000001\n\nuniform float gamma1;\nuniform float gamma2;\nuniform float gamma3;\nuniform float gamma4;\n\nfloat gammaContrast(float arr, float g) {\n  // Gamma must be > 0\n  g = clamp(g, epsilon, g);\n\n  return pow(arr, 1.0 / g);\n}\n\nvec4 gammaContrast(vec4 arr, float g1, float g2, float g3, float g4) {\n  arr.r = gammaContrast(arr.r, g1);\n  arr.g = gammaContrast(arr.g, g2);\n  arr.b = gammaContrast(arr.b, g3);\n  arr.a = gammaContrast(arr.a, g4);\n\n  return arr;\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var gammaContrastValue = opts.gammaContrastValue,
    gammaContrastValue1 = opts.gammaContrastValue1,
    gammaContrastValue2 = opts.gammaContrastValue2,
    gammaContrastValue3 = opts.gammaContrastValue3,
    gammaContrastValue4 = opts.gammaContrastValue4;

  // Gamma must be > 0, so not using Number.isFinite is fine

  if (gammaContrastValue) {
    return {
      gamma1: gammaContrastValue,
      gamma2: gammaContrastValue,
      gamma3: gammaContrastValue,
      gamma4: gammaContrastValue
    };
  } else if (gammaContrastValue1 || gammaContrastValue2 || gammaContrastValue3 || gammaContrastValue4) {
    return {
      gamma1: gammaContrastValue1 || 1,
      gamma2: gammaContrastValue2 || 1,
      gamma3: gammaContrastValue3 || 1,
      gamma4: gammaContrastValue4 || 1
    };
  }
  return null;
}
var gammaContrast = exports.gammaContrast = {
  name: 'gamma_contrast',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    image = gammaContrast(image, gamma1, gamma2, gamma3, gamma4);\n    "
  }
};
