// @kepler.gl/deckgl-layers/dist/raster/webgl/color/linear-rescale.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.linearRescale = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var fs = "uniform float linearRescaleScaler;\nuniform float linearRescaleOffset;\n\n// Perform a linear rescaling of image\nvec4 linear_rescale(vec4 arr, float scaler, float offset) {\n  return arr * scaler + offset;\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var linearRescaleScaler = opts.linearRescaleScaler,
    linearRescaleOffset = opts.linearRescaleOffset;
  if (!Number.isFinite(linearRescaleScaler) && !Number.isFinite(linearRescaleOffset)) {
    return null;
  }
  return {
    linearRescaleScaler: Number.isFinite(linearRescaleScaler) ? linearRescaleScaler : 1,
    linearRescaleOffset: Number.isFinite(linearRescaleOffset) ? linearRescaleOffset : 0
  };
}
var linearRescale = exports.linearRescale = {
  name: 'linear_rescale',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    image = linear_rescale(image, linearRescaleScaler, linearRescaleOffset);\n    "
  }
};
