// @kepler.gl/deckgl-layers/dist/raster/webgl/color/saturation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.saturation = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

/**
 * Adjusts the saturation of a color.
 * From cesium:
 * https://github.com/CesiumGS/cesium/blob/master/Source/Shaders/Builtin/Functions/saturation.glsl
 *
 * @param {vec3} rgb The color.
 * @param {float} adjustment The amount to adjust the saturation of the color. Usually between 0 and 2.
 *
 * @returns {vec3} The color with the saturation adjusted.
 */
var fs = "uniform float uSaturationValue;\nvec3 saturate(vec3 rgb, float adjustment) {\n    // Algorithm from Chapter 16 of OpenGL Shading Language\n    const vec3 W = vec3(0.2125, 0.7154, 0.0721);\n    vec3 intensity = vec3(dot(rgb, W));\n    return mix(intensity, rgb, adjustment);\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var saturationValue = opts.saturationValue;
  if (!saturationValue) {
    return null;
  }
  return {
    uSaturationValue: Number.isFinite(saturationValue) ? saturationValue : 1
  };
}
var saturation = exports.saturation = {
  name: 'saturation',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    image = vec4(saturate(image.rgb, uSaturationValue), image.a);\n    "
  }
};
