// @kepler.gl/deckgl-layers/dist/raster/webgl/color/sigmoidal-contrast.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sigmoidalContrast = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

// From mapbox/rio-color under the MIT License
//
// Sigmoidal contrast is a type of contrast control that
// adjusts the contrast without saturating highlights or shadows.
// It allows control over two factors:
// the contrast range from light to dark, and where the middle value
// of the mid-tones falls. The result is a non-linear and smooth
// contrast change.
// Parameters
// ----------
// arr : ndarray, float, 0 .. 1
//     Array of color values to adjust
// contrast : integer
//     Enhances the intensity differences between the lighter and darker
//     elements of the image. For example, 0 is none, 3 is typical and
//     20 is a lot.
// bias : float, between 0 and 1
//     Threshold level for the contrast function to center on
//     (typically centered at 0.5)
// Notes
// ----------
// Sigmoidal contrast is based on the sigmoidal transfer function:
// .. math:: g(u) = ( 1/(1 + e^{- \alpha * u + \beta)})
// This sigmoid function is scaled so that the output is bound by
// the interval [0, 1].
// .. math:: ( 1/(1 + e^(\beta * (\alpha - u))) - 1/(1 + e^(\beta * \alpha)))/
//     ( 1/(1 + e^(\beta*(\alpha - 1))) - 1/(1 + e^(\beta * \alpha)) )
// Where :math: `\alpha` is the threshold level, and :math: `\beta` the
// contrast factor to be applied.
// References
// ----------
// .. [CT] Hany Farid "Fundamentals of Image Processing"
//         http://www.cs.dartmouth.edu/farid/downloads/tutorials/fip.pdf
var fs = "#define epsilon 0.00000001\n\nuniform float sigmoidalContrast;\nuniform float sigmoidalBias;\n\n// NOTE: Input array must have float values between 0 and 1!\n// NOTE: bias must be a scalar float between 0 and 1!\nvec4 calculateSigmoidalContrast(vec4 arr, float contrast, float bias) {\n  // We use the names alpha and beta to match documentation.\n  float alpha = bias;\n  float beta = contrast;\n\n  // alpha must be >= 0\n  alpha = clamp(alpha, epsilon, alpha);\n\n  if (beta > 0.) {\n    vec4 numerator = 1. / (1. + exp(beta * (alpha - arr))) - 1. / (\n      1. + exp(beta * alpha)\n    );\n    float denominator = 1. / (1. + exp(beta * (alpha - 1.))) - 1. / (\n      1. + exp(beta * alpha)\n    );\n    arr = numerator / denominator;\n  } else if (beta < 0.) {\n    arr = (\n      (beta * alpha) - log(\n        (\n          1.0 / (\n            (arr / (1.0 + exp((beta * alpha) - beta))) -\n            (arr / (1.0 + exp(beta * alpha))) +\n            (1.0 / (1.0 + exp(beta * alpha)))\n          )\n        ) - 1.0)\n    ) / beta;\n  }\n\n  return arr;\n}\n";
function getUniforms() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var sigmoidalContrast = opts.sigmoidalContrast,
    sigmoidalBias = opts.sigmoidalBias;
  if (!(Number.isFinite(sigmoidalContrast) || Number.isFinite(sigmoidalBias))) {
    return null;
  }
  return {
    sigmoidalContrast: Number.isFinite(sigmoidalContrast) ? sigmoidalContrast : 0,
    sigmoidalBias: Number.isFinite(sigmoidalBias) ? sigmoidalBias : 0.5
  };
}
var sigmoidalContrast = exports.sigmoidalContrast = {
  name: 'sigmoidalContrast',
  fs: fs,
  getUniforms: getUniforms,
  inject: {
    'fs:DECKGL_MUTATE_COLOR': "\n    image = calculateSigmoidalContrast(image, sigmoidalContrast, sigmoidalBias);\n    "
  }
};
