// @kepler.gl/reducers/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  keplerGlReducer: true,
  keplerGlReducerCore: true,
  visStateLens: true,
  mapStateLens: true,
  uiStateLens: true,
  mapStyleLens: true,
  visStateReducer: true,
  mapStateReducer: true,
  uiStateReducer: true,
  mapStyleReducer: true,
  providerReducer: true,
  visStateUpdaters: true,
  fitBoundsUpdater: true,
  pickViewportPropsFromMapState: true,
  INITIAL_MAP_STATE: true,
  mapStateUpdaters: true,
  getInitialInputStyle: true,
  loadMapStylesUpdater: true,
  INITIAL_MAP_STYLE: true,
  mapStyleUpdaters: true,
  INITIAL_UI_STATE: true,
  uiStateUpdaters: true,
  addDataToMapUpdater: true,
  replaceDataInMapUpdater: true,
  combineUpdaters: true,
  combinedUpdaters: true,
  visStateMergers: true,
  providerStateUpdaters: true,
  enhanceReduxMiddleware: true
};
Object.defineProperty(exports, "INITIAL_MAP_STATE", {
  enumerable: true,
  get: function get() {
    return _mapStateUpdaters.INITIAL_MAP_STATE;
  }
});
Object.defineProperty(exports, "INITIAL_MAP_STYLE", {
  enumerable: true,
  get: function get() {
    return _mapStyleUpdaters.INITIAL_MAP_STYLE;
  }
});
Object.defineProperty(exports, "INITIAL_UI_STATE", {
  enumerable: true,
  get: function get() {
    return _uiStateUpdaters.INITIAL_UI_STATE;
  }
});
Object.defineProperty(exports, "addDataToMapUpdater", {
  enumerable: true,
  get: function get() {
    return _combineUpdaters.addDataToMapUpdater;
  }
});
exports.combinedUpdaters = exports.combineUpdaters = void 0;
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _root["default"];
  }
});
Object.defineProperty(exports, "enhanceReduxMiddleware", {
  enumerable: true,
  get: function get() {
    return _middleware.enhanceReduxMiddleware;
  }
});
Object.defineProperty(exports, "fitBoundsUpdater", {
  enumerable: true,
  get: function get() {
    return _mapStateUpdaters.fitBoundsUpdater;
  }
});
Object.defineProperty(exports, "getInitialInputStyle", {
  enumerable: true,
  get: function get() {
    return _mapStyleUpdaters.getInitialInputStyle;
  }
});
Object.defineProperty(exports, "keplerGlReducer", {
  enumerable: true,
  get: function get() {
    return _root["default"];
  }
});
Object.defineProperty(exports, "keplerGlReducerCore", {
  enumerable: true,
  get: function get() {
    return _core["default"];
  }
});
Object.defineProperty(exports, "loadMapStylesUpdater", {
  enumerable: true,
  get: function get() {
    return _mapStyleUpdaters.loadMapStylesUpdater;
  }
});
Object.defineProperty(exports, "mapStateLens", {
  enumerable: true,
  get: function get() {
    return _core.mapStateLens;
  }
});
Object.defineProperty(exports, "mapStateReducer", {
  enumerable: true,
  get: function get() {
    return _mapState["default"];
  }
});
exports.mapStateUpdaters = void 0;
Object.defineProperty(exports, "mapStyleLens", {
  enumerable: true,
  get: function get() {
    return _core.mapStyleLens;
  }
});
Object.defineProperty(exports, "mapStyleReducer", {
  enumerable: true,
  get: function get() {
    return _mapStyle["default"];
  }
});
exports.mapStyleUpdaters = void 0;
Object.defineProperty(exports, "pickViewportPropsFromMapState", {
  enumerable: true,
  get: function get() {
    return _mapStateUpdaters.pickViewportPropsFromMapState;
  }
});
Object.defineProperty(exports, "providerReducer", {
  enumerable: true,
  get: function get() {
    return _providerState["default"];
  }
});
exports.providerStateUpdaters = void 0;
Object.defineProperty(exports, "replaceDataInMapUpdater", {
  enumerable: true,
  get: function get() {
    return _combineUpdaters.replaceDataInMapUpdater;
  }
});
Object.defineProperty(exports, "uiStateLens", {
  enumerable: true,
  get: function get() {
    return _core.uiStateLens;
  }
});
Object.defineProperty(exports, "uiStateReducer", {
  enumerable: true,
  get: function get() {
    return _uiState["default"];
  }
});
exports.uiStateUpdaters = void 0;
Object.defineProperty(exports, "visStateLens", {
  enumerable: true,
  get: function get() {
    return _core.visStateLens;
  }
});
exports.visStateMergers = void 0;
Object.defineProperty(exports, "visStateReducer", {
  enumerable: true,
  get: function get() {
    return _visState["default"];
  }
});
exports.visStateUpdaters = void 0;
var _root = _interopRequireDefault(require("./root"));
var _core = _interopRequireWildcard(require("./core"));
var _visState = _interopRequireDefault(require("./vis-state"));
var _mapState = _interopRequireWildcard(require("./map-state"));
Object.keys(_mapState).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapState[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _mapState[key];
    }
  });
});
var _uiState = _interopRequireWildcard(require("./ui-state"));
Object.keys(_uiState).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _uiState[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _uiState[key];
    }
  });
});
var _mapStyle = _interopRequireDefault(require("./map-style"));
var _providerState = _interopRequireWildcard(require("./provider-state"));
Object.keys(_providerState).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _providerState[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _providerState[key];
    }
  });
});
var _visStateUpdaters = _interopRequireWildcard(require("./vis-state-updaters"));
exports.visStateUpdaters = _visStateUpdaters;
Object.keys(_visStateUpdaters).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _visStateUpdaters[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _visStateUpdaters[key];
    }
  });
});
var _mapStateUpdaters = _interopRequireWildcard(require("./map-state-updaters"));
exports.mapStateUpdaters = _mapStateUpdaters;
var _mapStyleUpdaters = _interopRequireWildcard(require("./map-style-updaters"));
exports.mapStyleUpdaters = _mapStyleUpdaters;
var _uiStateUpdaters = _interopRequireWildcard(require("./ui-state-updaters"));
exports.uiStateUpdaters = _uiStateUpdaters;
var _combineUpdaters = _interopRequireWildcard(require("./combined-updaters"));
var _combinedUpdaters = _combineUpdaters;
exports.combineUpdaters = _combineUpdaters;
exports.combinedUpdaters = _combineUpdaters;
var _visStateMergers = _interopRequireWildcard(require("./vis-state-merger"));
exports.visStateMergers = _visStateMergers;
Object.keys(_visStateMergers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _visStateMergers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _visStateMergers[key];
    }
  });
});
var _visStateSelectors = require("./vis-state-selectors");
Object.keys(_visStateSelectors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _visStateSelectors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _visStateSelectors[key];
    }
  });
});
var _providerStateUpdaters = _interopRequireWildcard(require("./provider-state-updaters"));
exports.providerStateUpdaters = _providerStateUpdaters;
Object.keys(_providerStateUpdaters).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _providerStateUpdaters[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _providerStateUpdaters[key];
    }
  });
});
var _composerHelpers = require("./composer-helpers");
Object.keys(_composerHelpers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _composerHelpers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _composerHelpers[key];
    }
  });
});
var _dataUtils = require("./data-utils");
Object.keys(_dataUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dataUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _dataUtils[key];
    }
  });
});
var _exportUtils = require("./export-utils");
Object.keys(_exportUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _exportUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _exportUtils[key];
    }
  });
});
var _interactionUtils = require("./interaction-utils");
Object.keys(_interactionUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interactionUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _interactionUtils[key];
    }
  });
});
var _layerUtils = require("./layer-utils");
Object.keys(_layerUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _layerUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _layerUtils[key];
    }
  });
});
var _middleware = require("./middleware");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
