// @kepler.gl/layers/dist/mapboxgl-layer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pointColResolver = exports.mapboxRequiredColumns = exports["default"] = exports.COLUMN_MODE_POINTS = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _baseLayer = _interopRequireWildcard(require("./base-layer"));
var _reselect = require("reselect");
var _mapboxUtils = require("./mapbox-utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2["default"])(o), (0, _possibleConstructorReturn2["default"])(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2["default"])(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var COLUMN_MODE_POINTS = exports.COLUMN_MODE_POINTS = 'points';
var mapboxRequiredColumns = exports.mapboxRequiredColumns = ['lat', 'lng'];
var SUPPORTED_COLUMN_MODES = [{
  key: COLUMN_MODE_POINTS,
  label: 'Points',
  requiredColumns: mapboxRequiredColumns
}];
var pointColResolver = exports.pointColResolver = function pointColResolver(_ref, columnMode) {
  var lat = _ref.lat,
    lng = _ref.lng,
    geoarrow = _ref.geoarrow;
  return "".concat(columnMode, "-").concat(lat.fieldIdx, "-").concat(lng.fieldIdx, "-").concat(geoarrow === null || geoarrow === void 0 ? void 0 : geoarrow.fieldIdx);
};
var MapboxLayerGL = /*#__PURE__*/function (_Layer) {
  function MapboxLayerGL() {
    var _this;
    (0, _classCallCheck2["default"])(this, MapboxLayerGL);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, MapboxLayerGL, [].concat(args));
    (0, _defineProperty2["default"])(_this, "datasetSelector", function (config) {
      return config.dataId;
    });
    (0, _defineProperty2["default"])(_this, "gpuFilterSelector", function (config, datasets) {
      return (config.dataId && datasets[config.dataId] || {}).gpuFilter;
    });
    (0, _defineProperty2["default"])(_this, "columnsSelector", function (config) {
      return pointColResolver(config.columns, config.columnMode);
    });
    (0, _defineProperty2["default"])(_this, "sourceSelector", (0, _reselect.createSelector)(_this.datasetSelector, _this.columnsSelector, function (datasetId, columns) {
      return "".concat(datasetId, "-").concat(columns);
    }));
    (0, _defineProperty2["default"])(_this, "filterSelector", (0, _reselect.createSelector)(_this.gpuFilterSelector, function (gpuFilter) {
      return (0, _mapboxUtils.gpuFilterToMapboxFilter)(gpuFilter);
    }));
    return _this;
  }
  (0, _inherits2["default"])(MapboxLayerGL, _Layer);
  return (0, _createClass2["default"])(MapboxLayerGL, [{
    key: "overlayType",
    get: function get() {
      return _baseLayer.OVERLAY_TYPE_CONST.mapboxgl;
    }
  }, {
    key: "type",
    get: function get() {
      return null;
    }
  }, {
    key: "isAggregated",
    get: function get() {
      return true;
    }
  }, {
    key: "supportedColumnModes",
    get: function get() {
      return SUPPORTED_COLUMN_MODES;
    }
  }, {
    key: "columnPairs",
    get: function get() {
      return this.defaultPointColumnPairs;
    }
  }, {
    key: "noneLayerDataAffectingProps",
    get: function get() {
      return [];
    }
  }, {
    key: "visualChannels",
    get: function get() {
      return {};
    }
  }, {
    key: "isValidFilter",
    value: function isValidFilter(filter) {
      // mapbox will crash if filter is not an array or empty
      return Array.isArray(filter) && filter.length;
    }
  }, {
    key: "getDataUpdateTriggers",
    value: function getDataUpdateTriggers(_ref2) {
      var _this2 = this;
      var filteredIndex = _ref2.filteredIndex,
        gpuFilter = _ref2.gpuFilter,
        id = _ref2.id;
      var columns = this.config.columns;
      var visualChannelFields = Object.values(this.visualChannels).reduce(function (accu, v) {
        return _objectSpread(_objectSpread({}, accu), _this2.config[v.field] ? (0, _defineProperty2["default"])({}, v.field, _this2.config[v.field].name) : {});
      }, {});
      var updateTriggers = {
        getData: _objectSpread(_objectSpread({
          datasetId: id,
          columns: columns,
          filteredIndex: filteredIndex
        }, visualChannelFields), gpuFilter.filterValueUpdateTriggers),
        getMeta: {
          datasetId: id,
          columns: columns
        }
      };
      return updateTriggers;
    }
  }, {
    key: "getGeometry",
    value: function getGeometry(position) {
      return position.every(Number.isFinite) ? {
        type: 'Point',
        coordinates: position
      } : null;
    }
  }, {
    key: "calculateDataAttribute",
    value: function calculateDataAttribute(_ref4, getPosition) {
      var _this3 = this;
      var dataContainer = _ref4.dataContainer,
        filteredIndex = _ref4.filteredIndex,
        gpuFilter = _ref4.gpuFilter;
      var getGeometry = function getGeometry(d) {
        return _this3.getGeometry(getPosition(d));
      };
      var vcFields = Object.values(this.visualChannels).map(function (v) {
        return _this3.config[v.field];
      }).filter(function (v) {
        return v;
      });
      var getPropertyFromVisualChanel = vcFields.length ? function (d) {
        return vcFields.reduce(function (accu, field) {
          return _objectSpread(_objectSpread({}, accu), {}, (0, _defineProperty2["default"])({}, field.name, field.valueAccessor(d)));
        }, {});
      } : function () {
        return {};
      };
      var filterValueUpdateTriggers = gpuFilter.filterValueUpdateTriggers,
        filterValueAccessor = gpuFilter.filterValueAccessor;

      // gpuField To property
      var hasFilter = Object.values(filterValueUpdateTriggers).filter(function (d) {
        return d;
      }).length;
      var valueAccessor = filterValueAccessor(dataContainer)();
      var getPropertyFromFilter = hasFilter ? function (d) {
        var filterValue = valueAccessor(d);
        return Object.values(filterValueUpdateTriggers).reduce(function (accu, gpu, i) {
          return _objectSpread(_objectSpread({}, accu), gpu !== null && gpu !== void 0 && gpu.name ? (0, _defineProperty2["default"])({}, (0, _mapboxUtils.prefixGpuField)(gpu.name), filterValue[i]) : {});
        }, {});
      } : function () {
        return {};
      };
      var getProperties = function getProperties(d) {
        return _objectSpread(_objectSpread({}, getPropertyFromVisualChanel(d)), getPropertyFromFilter(d));
      };
      return (0, _mapboxUtils.geoJsonFromData)(filteredIndex, getGeometry, getProperties);
    }

    // this layer is rendered at mapbox level
    // todo: maybe need to find a better solution for this one
  }, {
    key: "shouldRenderLayer",
    value: function shouldRenderLayer() {
      return typeof this.type === 'string' && this.config.isVisible && this.hasAllColumns();
    }
  }]);
}(_baseLayer["default"]);
var _default = exports["default"] = MapboxLayerGL;
