// @kepler.gl/constants/dist/time.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.durationYear = exports.durationWeek = exports.durationSecond = exports.durationMonth = exports.durationMinute = exports.durationMillisecond = exports.durationHour = exports.durationDay = exports.WOW = exports.TIME_INTERVALS = exports.TIME_AGGREGATION = exports.TICK_INTERVALS = exports.PLOT_TYPE_OPTIONS = exports.INTERVAL_OPTIONS = exports.INTERVAL = exports.DURATIONS = exports.BINS_LARGE = exports.BINS = exports.ANIMATION_TYPE = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _keymirror = _interopRequireDefault(require("keymirror"));
var _d3Time = require("d3-time");
var _ = require("./");
var _TICK_INTERVALS; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
// TODO: TimeInterval should be an enum

var durationMillisecond = exports.durationMillisecond = 1;
var durationSecond = exports.durationSecond = 1000;
var durationMinute = exports.durationMinute = durationSecond * 60;
var durationHour = exports.durationHour = durationMinute * 60;
var durationDay = exports.durationDay = durationHour * 24;
var durationWeek = exports.durationWeek = durationDay * 7;
var durationMonth = exports.durationMonth = durationDay * 30;
var durationYear = exports.durationYear = durationDay * 365;

// moment.duration functions
var DURATIONS = exports.DURATIONS = {
  years: durationYear,
  months: durationMonth,
  weeks: durationWeek,
  days: durationDay,
  hours: durationHour,
  minutes: durationMinute,
  seconds: durationSecond,
  milliseconds: durationMillisecond
};

// interval key to d3 time interval function
var TIME_INTERVALS = exports.TIME_INTERVALS = {
  year: _d3Time.utcYear,
  month: _d3Time.utcMonth,
  week: _d3Time.utcWeek,
  day: _d3Time.utcDay,
  hour: _d3Time.utcHour,
  minute: _d3Time.utcMinute,
  second: _d3Time.utcSecond,
  millisecond: _d3Time.utcMillisecond
};
var INTERVAL = exports.INTERVAL = (0, _keymirror["default"])({
  '1-second': null,
  '5-second': null,
  '15-second': null,
  '30-second': null,
  '1-minute': null,
  '5-minute': null,
  '15-minute': null,
  '30-minute': null,
  '1-hour': null,
  '3-hour': null,
  '6-hour': null,
  '12-hour': null,
  '1-day': null,
  '2-day': null,
  '1-week': null,
  '1-month': null,
  '3-month': null,
  '1-year': null
});

// interval eligible for calculate week over week
var WOW = exports.WOW = (0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, INTERVAL['1-day'], 7), INTERVAL['1-week'], 1);
var INTERVAL_OPTIONS = exports.INTERVAL_OPTIONS = Object.keys(INTERVAL).map(function (id) {
  var _id$split = id.split('-'),
    _id$split2 = (0, _slicedToArray2["default"])(_id$split, 2),
    step = _id$split2[0],
    interval = _id$split2[1];
  // capitalizeFirstLetter(interval)
  return {
    id: id,
    label: "".concat(step, " ").concat(interval)
  };
});
var TICK_INTERVALS = exports.TICK_INTERVALS = (_TICK_INTERVALS = {}, (0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])(_TICK_INTERVALS, INTERVAL['1-millisecond'], {
  interval: 'millisecond',
  step: 1,
  duration: durationMillisecond
}), INTERVAL['1-second'], {
  interval: 'second',
  step: 1,
  duration: durationSecond
}), INTERVAL['5-second'], {
  interval: 'second',
  step: 5,
  duration: 5 * durationSecond
}), INTERVAL['15-second'], {
  interval: 'second',
  step: 15,
  duration: 15 * durationSecond
}), INTERVAL['30-second'], {
  interval: 'second',
  step: 30,
  duration: 30 * durationSecond
}), INTERVAL['1-minute'], {
  interval: 'minute',
  step: 1,
  duration: durationMinute
}), INTERVAL['5-minute'], {
  interval: 'minute',
  step: 5,
  duration: 5 * durationMinute
}), INTERVAL['15-minute'], {
  interval: 'minute',
  step: 15,
  duration: 15 * durationMinute
}), INTERVAL['30-minute'], {
  interval: 'minute',
  step: 30,
  duration: 30 * durationMinute
}), INTERVAL['1-hour'], {
  interval: 'hour',
  step: 1,
  duration: durationHour
}), (0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])((0, _defineProperty2["default"])(_TICK_INTERVALS, INTERVAL['3-hour'], {
  interval: 'hour',
  step: 3,
  duration: 3 * durationHour
}), INTERVAL['6-hour'], {
  interval: 'hour',
  step: 6,
  duration: 6 * durationHour
}), INTERVAL['12-hour'], {
  interval: 'hour',
  step: 12,
  duration: 12 * durationHour
}), INTERVAL['1-day'], {
  interval: 'day',
  step: 1,
  duration: durationDay
}), INTERVAL['2-day'], {
  interval: 'day',
  step: 2,
  duration: 2 * durationDay
}), INTERVAL['1-week'], {
  interval: 'week',
  step: 1,
  duration: durationWeek
}), INTERVAL['1-month'], {
  interval: 'month',
  step: 1,
  duration: durationMonth
}), INTERVAL['3-month'], {
  interval: 'month',
  step: 3,
  duration: 3 * durationMonth
}), INTERVAL['1-year'], {
  interval: 'year',
  step: 1,
  duration: durationYear
}));
var PLOT_TYPE_OPTIONS = exports.PLOT_TYPE_OPTIONS = (0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, _.PLOT_TYPES.lineChart, {
  id: _.PLOT_TYPES.lineChart,
  label: 'Chart',
  icon: 'LineChart'
}), _.PLOT_TYPES.histogram, {
  id: _.PLOT_TYPES.histogram,
  label: 'Histogram',
  icon: 'Histogram'
});
var BINS_LARGE = exports.BINS_LARGE = 100;
var BINS = exports.BINS = 30;
var TIME_AGGREGATION = exports.TIME_AGGREGATION = [{
  id: _.AGGREGATION_TYPES.average,
  label: 'Average'
}, {
  id: _.AGGREGATION_TYPES.sum,
  label: 'Sum'
}, {
  id: _.AGGREGATION_TYPES.maximum,
  label: 'Maximum'
}, {
  id: _.AGGREGATION_TYPES.minimum,
  label: 'Minimum'
}, {
  id: _.AGGREGATION_TYPES.median,
  label: 'Median'
}, {
  id: _.AGGREGATION_TYPES.stdev,
  label: 'Std Deviation'
}];
var ANIMATION_TYPE = exports.ANIMATION_TYPE = (0, _keymirror["default"])({
  interval: null,
  continuous: null
});
