// @kepler.gl/components/dist/common/logo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _constants = require("@kepler.gl/constants");
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var LogoTitle = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: inline-block;\n  margin-left: 6px;\n"])));
var LogoName = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  .logo__link {\n    color: ", ";\n    font-size: 14px;\n    font-weight: 600;\n    letter-spacing: 1.17px;\n  }\n"])), function (props) {
  return props.theme.logoColor;
});
var LogoVersion = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 10px;\n  color: ", ";\n  letter-spacing: 0.83px;\n  line-height: 14px;\n"])), function (props) {
  return props.theme.subtextColor;
});
var LogoWrapper = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-items: flex-start;\n"])));
var LogoSvgWrapper = _styledComponents["default"].div(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  margin-top: 3px;\n"])));
var LogoSvg = function LogoSvg() {
  return /*#__PURE__*/_react["default"].createElement("svg", {
    className: "side-panel-logo__logo",
    width: "22px",
    height: "15px",
    viewBox: "0 0 22 15"
  }, /*#__PURE__*/_react["default"].createElement("g", {
    transform: "translate(11, -3) rotate(45.000000)"
  }, /*#__PURE__*/_react["default"].createElement("rect", {
    fill: "#535C6C",
    x: "0",
    y: "5",
    width: "10",
    height: "10"
  }), /*#__PURE__*/_react["default"].createElement("rect", {
    fill: "#1FBAD6",
    x: "5",
    y: "0",
    width: "10",
    height: "10"
  })));
};
var KeplerGlLogo = function KeplerGlLogo(_ref) {
  var _ref$appName = _ref.appName,
    appName = _ref$appName === void 0 ? _constants.KEPLER_GL_NAME : _ref$appName,
    _ref$appWebsite = _ref.appWebsite,
    appWebsite = _ref$appWebsite === void 0 ? _constants.KEPLER_GL_WEBSITE : _ref$appWebsite,
    _ref$version = _ref.version,
    version = _ref$version === void 0 ? _constants.KEPLER_GL_VERSION : _ref$version;
  return /*#__PURE__*/_react["default"].createElement(LogoWrapper, {
    className: "side-panel-logo"
  }, /*#__PURE__*/_react["default"].createElement(LogoSvgWrapper, null, /*#__PURE__*/_react["default"].createElement(LogoSvg, null)), /*#__PURE__*/_react["default"].createElement(LogoTitle, {
    className: "logo__title"
  }, /*#__PURE__*/_react["default"].createElement(LogoName, {
    className: "logo__name"
  }, /*#__PURE__*/_react["default"].createElement("a", {
    className: "logo__link",
    target: "_blank",
    rel: "noopener noreferrer",
    href: appWebsite
  }, appName)), version ? /*#__PURE__*/_react["default"].createElement(LogoVersion, {
    className: "logo__version"
  }, version) : null));
};
var _default = exports["default"] = KeplerGlLogo;
