// @kepler.gl/components/dist/common/range-slider-timeline.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _icons = require("../common/icons");
var _styledComponents = require("../common/styled-components");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var BACKGROUND_LINE_STYLE = {
  height: '4px',
  backgroundColor: '#1C2233',
  position: 'relative',
  width: '100%',
  marginTop: '20px'
};
var TIMELINE_MARKER_STYLE = {
  position: 'absolute',
  top: '-8px',
  fill: '#3D4866',
  color: '#3D4866'
};
var containerStyle = {
  display: 'flex',
  width: '100%',
  height: '24px'
};
var iconWrapperStyle = {
  marginRight: '8px',
  cursor: 'pointer'
};
var TIMELINE_INDICATOR_STYLE = {
  position: 'absolute',
  top: '-14px',
  fill: '#C4C4C4',
  color: '#C4C4C4'
};
function RangeSliderTimelineFactory() {
  var RangeSliderTimeline = function RangeSliderTimeline(_ref) {
    var subAnimation = _ref.subAnimation,
      scaledValue = _ref.scaledValue,
      style = _ref.style;
    var startTime = subAnimation.startTime,
      endTime = subAnimation.endTime,
      syncMode = subAnimation.syncMode,
      Icon = subAnimation.Icon,
      label = subAnimation.label;
    var progressStyle = {
      left: "".concat(startTime, "%"),
      top: '0',
      width: "".concat(endTime - startTime, "%"),
      height: '100%',
      position: 'absolute',
      backgroundColor: '#5558DB'
    };
    var progressBarContainer = (0, _react.useMemo)(function () {
      return _objectSpread(_objectSpread({}, BACKGROUND_LINE_STYLE), {}, {
        flex: 1
      }, style);
    }, [style]);
    var value = scaledValue[syncMode];
    var leftMarketStyle = (0, _react.useMemo)(function () {
      return _objectSpread({
        left: "calc(".concat(startTime, "% - 4px)")
      }, TIMELINE_MARKER_STYLE);
    }, [startTime]);
    var rightMarketStyle = (0, _react.useMemo)(function () {
      return _objectSpread({
        left: "calc(".concat(endTime, "% - 4px)")
      }, TIMELINE_MARKER_STYLE);
    }, [endTime]);
    var indicatorStyle = (0, _react.useMemo)(function () {
      return _objectSpread(_objectSpread({}, TIMELINE_INDICATOR_STYLE), {}, {
        left: "calc(".concat(value, "% - 3px)")
      });
    }, [value]);
    return /*#__PURE__*/_react["default"].createElement("div", {
      style: containerStyle
    }, /*#__PURE__*/_react["default"].createElement("div", {
      "data-tip": true,
      "data-for": label,
      style: iconWrapperStyle
    }, /*#__PURE__*/_react["default"].createElement(Icon, {
      height: "24px",
      color: "#F7F8FA"
    }), /*#__PURE__*/_react["default"].createElement(_styledComponents.Tooltip, {
      id: label,
      place: "right",
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement("span", null, label))), /*#__PURE__*/_react["default"].createElement("div", {
      style: progressBarContainer
    }, /*#__PURE__*/_react["default"].createElement("div", {
      style: progressStyle
    }), /*#__PURE__*/_react["default"].createElement(_icons.ArrowDownFull, {
      style: leftMarketStyle
    }), /*#__PURE__*/_react["default"].createElement(_icons.ArrowDownFull, {
      style: rightMarketStyle
    }), /*#__PURE__*/_react["default"].createElement(_icons.TimelineMarker, {
      style: indicatorStyle
    })));
  };
  return RangeSliderTimeline;
}
var _default = exports["default"] = RangeSliderTimelineFactory;
