// @kepler.gl/components/dist/common/animation-control/reset-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _localization = require("@kepler.gl/localization");
var _iconButton = _interopRequireDefault(require("../icon-button"));
var _tippyTooltip = _interopRequireDefault(require("../tippy-tooltip"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

function ResetControlFactory() {
  var ResetControl = function ResetControl(_ref) {
    var showAnimationWindowControl = _ref.showAnimationWindowControl,
      resetAnimation = _ref.resetAnimation,
      btnStyle = _ref.btnStyle,
      playbackIcons = _ref.playbackIcons,
      buttonHeight = _ref.buttonHeight;
    return showAnimationWindowControl ? null : /*#__PURE__*/_react["default"].createElement(_tippyTooltip["default"], {
      placement: "top",
      delay: [500, 0],
      render: function render() {
        return /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
          id: "tooltip.reset"
        });
      }
    }, /*#__PURE__*/_react["default"].createElement(_iconButton["default"], (0, _extends2["default"])({
      className: "playback-control-button",
      onClick: resetAnimation
    }, btnStyle), /*#__PURE__*/_react["default"].createElement(playbackIcons.reset, {
      height: buttonHeight
    })));
  };
  return ResetControl;
}
var _default = exports["default"] = ResetControlFactory;
