// @kepler.gl/components/dist/common/animation-control/speed-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _iconButton = _interopRequireDefault(require("../icon-button"));
var _styles = require("@kepler.gl/styles");
var _utils = require("@kepler.gl/utils");
var _tippyTooltip = _interopRequireDefault(require("../tippy-tooltip"));
var _templateObject, _templateObject2; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var StyledSpeedControl = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-items: center;\n\n  .animation-control__speed-slider {\n    left: 0;\n    ", "\n  }\n"])), _styles.media.palm(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n      left: 60px;\n    "]))));
var PRECISE_SPEED_ROUND = 1;
function SpeedControlFactory(AnimationSpeedSlider) {
  var SpeedControl = function SpeedControl(_ref) {
    var showAnimationWindowControl = _ref.showAnimationWindowControl,
      updateAnimationSpeed = _ref.updateAnimationSpeed,
      btnStyle = _ref.btnStyle,
      hideAndShowSpeedControl = _ref.hideAndShowSpeedControl,
      buttonHeight = _ref.buttonHeight,
      playbackIcons = _ref.playbackIcons,
      speed = _ref.speed,
      isSpeedControlVisible = _ref.isSpeedControlVisible;
    return showAnimationWindowControl || !updateAnimationSpeed ? null : /*#__PURE__*/_react["default"].createElement(StyledSpeedControl, null, /*#__PURE__*/_react["default"].createElement(_tippyTooltip["default"], {
      placement: "top",
      delay: [500, 0],
      render: function render() {
        return /*#__PURE__*/_react["default"].createElement("span", null, (0, _utils.preciseRound)(speed, PRECISE_SPEED_ROUND), "x");
      }
    }, /*#__PURE__*/_react["default"].createElement(_iconButton["default"], (0, _extends2["default"])({
      className: "playback-control-button"
    }, btnStyle, {
      onClick: hideAndShowSpeedControl
    }), /*#__PURE__*/_react["default"].createElement(playbackIcons.speed, {
      height: buttonHeight
    }))), isSpeedControlVisible ? /*#__PURE__*/_react["default"].createElement(AnimationSpeedSlider, {
      onHide: hideAndShowSpeedControl,
      updateAnimationSpeed: updateAnimationSpeed,
      speed: speed
    }) : null);
  };
  return SpeedControl;
}
var _default = exports["default"] = SpeedControlFactory;
