// @kepler.gl/utils/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  colorMaybeToRGB: true,
  colorRangeBackwardCompatibility: true,
  createLinearGradient: true,
  hasColorMap: true,
  hexToRgb: true,
  isHexColor: true,
  isRgbColor: true,
  normalizeColor: true,
  reverseColorRange: true,
  rgbToHex: true,
  addCustomPaletteColor: true,
  removeCustomPaletteColor: true,
  sortCustomPaletteColor: true,
  updateCustomPaletteColor: true,
  updateColorRangeBySelectedPalette: true,
  paletteIsSteps: true,
  paletteIsType: true,
  paletteIsColorBlindSafe: true,
  updateColorRangeByMatchingPalette: true,
  updateCustomColorRangeByColorUI: true,
  initializeCustomPalette: true,
  errorNotification: true,
  createNotification: true,
  exportImageError: true,
  successNotification: true,
  setStyleSheetBaseHref: true,
  domtoimage: true,
  getFrequency: true,
  getMode: true,
  aggregate: true,
  adjustValueToAnimationWindow: true,
  getBinThresholds: true,
  histogramFromThreshold: true,
  histogramFromValues: true,
  histogramFromDomain: true,
  histogramFromOrdinal: true,
  runGpuFilterForPlot: true,
  updateTimeFilterPlotType: true,
  snapToMarks: true,
  getTimeBins: true,
  SAMPLE_TIMELINE: true,
  TIMELINE_MODES: true,
  TIME_INTERVALS_ORDERED: true,
  LayerToFilterTimeInterval: true,
  TileTimeInterval: true,
  getTimelineFromAnimationConfig: true,
  getTimelineFromFilter: true,
  maybeHexToGeo: true,
  getPositionFromHexValue: true,
  datasetColorMaker: true,
  findDefaultColorField: true,
  getFieldFormatLabels: true,
  getFormatLabels: true,
  validateInputData: true,
  exportMapToHTML: true,
  calculateExportImageSize: true,
  convertToPng: true,
  dataURItoBlob: true,
  downloadFile: true,
  exportHtml: true,
  exportImage: true,
  exportJson: true,
  exportMap: true,
  exportToJsonString: true,
  exporters: true,
  getMapJSON: true,
  getScaleFromImageSize: true,
  isMSEdge: true,
  getFormatValue: true,
  getDefaultTimeFormat: true,
  setLayerBlending: true,
  flattenMessages: true,
  mergeMessages: true,
  isValidMapInfo: true,
  editBottomMapStyle: true,
  editTopMapStyle: true,
  getDefaultLayerGroupVisibility: true,
  getStyleDownloadUrl: true,
  getStyleImageIcon: true,
  mergeLayerGroupVisibility: true,
  scaleMapStyleByResolution: true,
  validateToken: true,
  useDimensions: true,
  observeDimensions: true,
  unobserveDimensions: true,
  computeDeckEffects: true,
  fixEffectOrder: true,
  reorderEffectOrder: true,
  validateEffectParameters: true,
  isStyleUsingMapboxTiles: true,
  isStyleUsingOpenStreetMapTiles: true,
  getBaseMapLibrary: true,
  transformRequest: true,
  ArrowDataContainer: true,
  arrowDataTypeToAnalyzerDataType: true,
  arrowDataTypeToFieldType: true,
  DataForm: true,
  createDataContainer: true,
  createIndexedDataContainer: true,
  getSampleContainerData: true,
  colorMapToColorBreaks: true,
  colorBreaksToColorMap: true,
  getLayerColorScale: true,
  getLegendOfScale: true,
  getLinearDomain: true,
  getLogDomain: true,
  getOrdinalDomain: true,
  getQuantileDomain: true,
  getScaleFunction: true,
  getVisualChannelScaleByZoom: true,
  initializeLayerColorMap: true,
  isNumericColorBreaks: true,
  isDomainStops: true,
  isDomainQuantile: true,
  getDomainStepsbyZoom: true,
  getThresholdsFromQuantiles: true,
  getQuantLabelFormat: true,
  getHistogramDomain: true,
  getQuantLegends: true,
  getCategoricalColorMap: true,
  getCategoricalColorScale: true,
  initCustomPaletteByCustomScale: true,
  colorMapToCategoricalColorBreaks: true,
  resetCategoricalColorMapByIndex: true,
  selectRestCategoricalColorMapByIndex: true,
  removeCategoricalValueFromColorMap: true,
  addCategoricalValuesToColorMap: true,
  DataRow: true,
  getApplicationConfig: true,
  initApplicationConfig: true,
  isAppleDevice: true,
  quickInsertionSort: true
};
Object.defineProperty(exports, "ArrowDataContainer", {
  enumerable: true,
  get: function get() {
    return _arrowDataContainer.ArrowDataContainer;
  }
});
Object.defineProperty(exports, "DataForm", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.DataForm;
  }
});
Object.defineProperty(exports, "DataRow", {
  enumerable: true,
  get: function get() {
    return _dataRow.DataRow;
  }
});
Object.defineProperty(exports, "LayerToFilterTimeInterval", {
  enumerable: true,
  get: function get() {
    return _time.LayerToFilterTimeInterval;
  }
});
Object.defineProperty(exports, "SAMPLE_TIMELINE", {
  enumerable: true,
  get: function get() {
    return _time.SAMPLE_TIMELINE;
  }
});
Object.defineProperty(exports, "TIMELINE_MODES", {
  enumerable: true,
  get: function get() {
    return _time.TIMELINE_MODES;
  }
});
Object.defineProperty(exports, "TIME_INTERVALS_ORDERED", {
  enumerable: true,
  get: function get() {
    return _time.TIME_INTERVALS_ORDERED;
  }
});
Object.defineProperty(exports, "TileTimeInterval", {
  enumerable: true,
  get: function get() {
    return _time.TileTimeInterval;
  }
});
Object.defineProperty(exports, "addCategoricalValuesToColorMap", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.addCategoricalValuesToColorMap;
  }
});
Object.defineProperty(exports, "addCustomPaletteColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.addCustomPaletteColor;
  }
});
Object.defineProperty(exports, "adjustValueToAnimationWindow", {
  enumerable: true,
  get: function get() {
    return _plot.adjustValueToAnimationWindow;
  }
});
Object.defineProperty(exports, "aggregate", {
  enumerable: true,
  get: function get() {
    return _aggregation.aggregate;
  }
});
Object.defineProperty(exports, "arrowDataTypeToAnalyzerDataType", {
  enumerable: true,
  get: function get() {
    return _arrowDataContainer.arrowDataTypeToAnalyzerDataType;
  }
});
Object.defineProperty(exports, "arrowDataTypeToFieldType", {
  enumerable: true,
  get: function get() {
    return _arrowDataContainer.arrowDataTypeToFieldType;
  }
});
Object.defineProperty(exports, "calculateExportImageSize", {
  enumerable: true,
  get: function get() {
    return _exportUtils.calculateExportImageSize;
  }
});
Object.defineProperty(exports, "colorBreaksToColorMap", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.colorBreaksToColorMap;
  }
});
Object.defineProperty(exports, "colorMapToCategoricalColorBreaks", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.colorMapToCategoricalColorBreaks;
  }
});
Object.defineProperty(exports, "colorMapToColorBreaks", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.colorMapToColorBreaks;
  }
});
Object.defineProperty(exports, "colorMaybeToRGB", {
  enumerable: true,
  get: function get() {
    return _colorUtils.colorMaybeToRGB;
  }
});
Object.defineProperty(exports, "colorRangeBackwardCompatibility", {
  enumerable: true,
  get: function get() {
    return _colorUtils.colorRangeBackwardCompatibility;
  }
});
Object.defineProperty(exports, "computeDeckEffects", {
  enumerable: true,
  get: function get() {
    return _effectUtils.computeDeckEffects;
  }
});
Object.defineProperty(exports, "convertToPng", {
  enumerable: true,
  get: function get() {
    return _exportUtils.convertToPng;
  }
});
Object.defineProperty(exports, "createDataContainer", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.createDataContainer;
  }
});
Object.defineProperty(exports, "createIndexedDataContainer", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.createIndexedDataContainer;
  }
});
Object.defineProperty(exports, "createLinearGradient", {
  enumerable: true,
  get: function get() {
    return _colorUtils.createLinearGradient;
  }
});
Object.defineProperty(exports, "createNotification", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.createNotification;
  }
});
Object.defineProperty(exports, "dataURItoBlob", {
  enumerable: true,
  get: function get() {
    return _exportUtils.dataURItoBlob;
  }
});
Object.defineProperty(exports, "datasetColorMaker", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.datasetColorMaker;
  }
});
Object.defineProperty(exports, "domtoimage", {
  enumerable: true,
  get: function get() {
    return _domToImage["default"];
  }
});
Object.defineProperty(exports, "downloadFile", {
  enumerable: true,
  get: function get() {
    return _exportUtils.downloadFile;
  }
});
Object.defineProperty(exports, "editBottomMapStyle", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.editBottomMapStyle;
  }
});
Object.defineProperty(exports, "editTopMapStyle", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.editTopMapStyle;
  }
});
Object.defineProperty(exports, "errorNotification", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.errorNotification;
  }
});
Object.defineProperty(exports, "exportHtml", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportHtml;
  }
});
Object.defineProperty(exports, "exportImage", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportImage;
  }
});
Object.defineProperty(exports, "exportImageError", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.exportImageError;
  }
});
Object.defineProperty(exports, "exportJson", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportJson;
  }
});
Object.defineProperty(exports, "exportMap", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportMap;
  }
});
Object.defineProperty(exports, "exportMapToHTML", {
  enumerable: true,
  get: function get() {
    return _exportMapHtml.exportMapToHTML;
  }
});
Object.defineProperty(exports, "exportToJsonString", {
  enumerable: true,
  get: function get() {
    return _exportUtils.exportToJsonString;
  }
});
Object.defineProperty(exports, "exporters", {
  enumerable: true,
  get: function get() {
    return _exportUtils["default"];
  }
});
Object.defineProperty(exports, "findDefaultColorField", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.findDefaultColorField;
  }
});
Object.defineProperty(exports, "fixEffectOrder", {
  enumerable: true,
  get: function get() {
    return _effectUtils.fixEffectOrder;
  }
});
Object.defineProperty(exports, "flattenMessages", {
  enumerable: true,
  get: function get() {
    return _localeUtils.flattenMessages;
  }
});
Object.defineProperty(exports, "getApplicationConfig", {
  enumerable: true,
  get: function get() {
    return _applicationConfig.getApplicationConfig;
  }
});
Object.defineProperty(exports, "getBaseMapLibrary", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils2.getBaseMapLibrary;
  }
});
Object.defineProperty(exports, "getBinThresholds", {
  enumerable: true,
  get: function get() {
    return _plot.getBinThresholds;
  }
});
Object.defineProperty(exports, "getCategoricalColorMap", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getCategoricalColorMap;
  }
});
Object.defineProperty(exports, "getCategoricalColorScale", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getCategoricalColorScale;
  }
});
Object.defineProperty(exports, "getDefaultLayerGroupVisibility", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.getDefaultLayerGroupVisibility;
  }
});
Object.defineProperty(exports, "getDefaultTimeFormat", {
  enumerable: true,
  get: function get() {
    return _format.getDefaultTimeFormat;
  }
});
Object.defineProperty(exports, "getDomainStepsbyZoom", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getDomainStepsbyZoom;
  }
});
Object.defineProperty(exports, "getFieldFormatLabels", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.getFieldFormatLabels;
  }
});
Object.defineProperty(exports, "getFormatLabels", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.getFormatLabels;
  }
});
Object.defineProperty(exports, "getFormatValue", {
  enumerable: true,
  get: function get() {
    return _format.getFormatValue;
  }
});
Object.defineProperty(exports, "getFrequency", {
  enumerable: true,
  get: function get() {
    return _aggregation.getFrequency;
  }
});
Object.defineProperty(exports, "getHistogramDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getHistogramDomain;
  }
});
Object.defineProperty(exports, "getLayerColorScale", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getLayerColorScale;
  }
});
Object.defineProperty(exports, "getLegendOfScale", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getLegendOfScale;
  }
});
Object.defineProperty(exports, "getLinearDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getLinearDomain;
  }
});
Object.defineProperty(exports, "getLogDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getLogDomain;
  }
});
Object.defineProperty(exports, "getMapJSON", {
  enumerable: true,
  get: function get() {
    return _exportUtils.getMapJSON;
  }
});
Object.defineProperty(exports, "getMode", {
  enumerable: true,
  get: function get() {
    return _aggregation.getMode;
  }
});
Object.defineProperty(exports, "getOrdinalDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getOrdinalDomain;
  }
});
Object.defineProperty(exports, "getPositionFromHexValue", {
  enumerable: true,
  get: function get() {
    return _positionUtils.getPositionFromHexValue;
  }
});
Object.defineProperty(exports, "getQuantLabelFormat", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getQuantLabelFormat;
  }
});
Object.defineProperty(exports, "getQuantLegends", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getQuantLegends;
  }
});
Object.defineProperty(exports, "getQuantileDomain", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getQuantileDomain;
  }
});
Object.defineProperty(exports, "getSampleContainerData", {
  enumerable: true,
  get: function get() {
    return _dataContainerUtils.getSampleData;
  }
});
Object.defineProperty(exports, "getScaleFromImageSize", {
  enumerable: true,
  get: function get() {
    return _exportUtils.getScaleFromImageSize;
  }
});
Object.defineProperty(exports, "getScaleFunction", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getScaleFunction;
  }
});
Object.defineProperty(exports, "getStyleDownloadUrl", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.getStyleDownloadUrl;
  }
});
Object.defineProperty(exports, "getStyleImageIcon", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.getStyleImageIcon;
  }
});
Object.defineProperty(exports, "getThresholdsFromQuantiles", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getThresholdsFromQuantiles;
  }
});
Object.defineProperty(exports, "getTimeBins", {
  enumerable: true,
  get: function get() {
    return _plot.getTimeBins;
  }
});
Object.defineProperty(exports, "getTimelineFromAnimationConfig", {
  enumerable: true,
  get: function get() {
    return _time.getTimelineFromAnimationConfig;
  }
});
Object.defineProperty(exports, "getTimelineFromFilter", {
  enumerable: true,
  get: function get() {
    return _time.getTimelineFromFilter;
  }
});
Object.defineProperty(exports, "getVisualChannelScaleByZoom", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.getVisualChannelScaleByZoom;
  }
});
Object.defineProperty(exports, "hasColorMap", {
  enumerable: true,
  get: function get() {
    return _colorUtils.hasColorMap;
  }
});
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _colorUtils.hexToRgb;
  }
});
Object.defineProperty(exports, "histogramFromDomain", {
  enumerable: true,
  get: function get() {
    return _plot.histogramFromDomain;
  }
});
Object.defineProperty(exports, "histogramFromOrdinal", {
  enumerable: true,
  get: function get() {
    return _plot.histogramFromOrdinal;
  }
});
Object.defineProperty(exports, "histogramFromThreshold", {
  enumerable: true,
  get: function get() {
    return _plot.histogramFromThreshold;
  }
});
Object.defineProperty(exports, "histogramFromValues", {
  enumerable: true,
  get: function get() {
    return _plot.histogramFromValues;
  }
});
Object.defineProperty(exports, "initApplicationConfig", {
  enumerable: true,
  get: function get() {
    return _applicationConfig.initApplicationConfig;
  }
});
Object.defineProperty(exports, "initCustomPaletteByCustomScale", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.initCustomPaletteByCustomScale;
  }
});
Object.defineProperty(exports, "initializeCustomPalette", {
  enumerable: true,
  get: function get() {
    return _colorUtils.initializeCustomPalette;
  }
});
Object.defineProperty(exports, "initializeLayerColorMap", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.initializeLayerColorMap;
  }
});
Object.defineProperty(exports, "isAppleDevice", {
  enumerable: true,
  get: function get() {
    return _browserUtils.isAppleDevice;
  }
});
Object.defineProperty(exports, "isDomainQuantile", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.isDomainQuantile;
  }
});
Object.defineProperty(exports, "isDomainStops", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.isDomainStops;
  }
});
Object.defineProperty(exports, "isHexColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.isHexColor;
  }
});
Object.defineProperty(exports, "isMSEdge", {
  enumerable: true,
  get: function get() {
    return _exportUtils.isMSEdge;
  }
});
Object.defineProperty(exports, "isNumericColorBreaks", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.isNumericColorBreaks;
  }
});
Object.defineProperty(exports, "isRgbColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.isRgbColor;
  }
});
Object.defineProperty(exports, "isStyleUsingMapboxTiles", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils2.isStyleUsingMapboxTiles;
  }
});
Object.defineProperty(exports, "isStyleUsingOpenStreetMapTiles", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils2.isStyleUsingOpenStreetMapTiles;
  }
});
Object.defineProperty(exports, "isValidMapInfo", {
  enumerable: true,
  get: function get() {
    return _mapInfoUtils.isValidMapInfo;
  }
});
Object.defineProperty(exports, "maybeHexToGeo", {
  enumerable: true,
  get: function get() {
    return _positionUtils.maybeHexToGeo;
  }
});
Object.defineProperty(exports, "mergeLayerGroupVisibility", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.mergeLayerGroupVisibility;
  }
});
Object.defineProperty(exports, "mergeMessages", {
  enumerable: true,
  get: function get() {
    return _localeUtils.mergeMessages;
  }
});
Object.defineProperty(exports, "normalizeColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.normalizeColor;
  }
});
Object.defineProperty(exports, "observeDimensions", {
  enumerable: true,
  get: function get() {
    return _observeDimensions.observeDimensions;
  }
});
Object.defineProperty(exports, "paletteIsColorBlindSafe", {
  enumerable: true,
  get: function get() {
    return _colorUtils.paletteIsColorBlindSafe;
  }
});
Object.defineProperty(exports, "paletteIsSteps", {
  enumerable: true,
  get: function get() {
    return _colorUtils.paletteIsSteps;
  }
});
Object.defineProperty(exports, "paletteIsType", {
  enumerable: true,
  get: function get() {
    return _colorUtils.paletteIsType;
  }
});
Object.defineProperty(exports, "quickInsertionSort", {
  enumerable: true,
  get: function get() {
    return _quickInsertionSort["default"];
  }
});
Object.defineProperty(exports, "removeCategoricalValueFromColorMap", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.removeCategoricalValueFromColorMap;
  }
});
Object.defineProperty(exports, "removeCustomPaletteColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.removeCustomPaletteColor;
  }
});
Object.defineProperty(exports, "reorderEffectOrder", {
  enumerable: true,
  get: function get() {
    return _effectUtils.reorderEffectOrder;
  }
});
Object.defineProperty(exports, "resetCategoricalColorMapByIndex", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.resetCategoricalColorMapByIndex;
  }
});
Object.defineProperty(exports, "reverseColorRange", {
  enumerable: true,
  get: function get() {
    return _colorUtils.reverseColorRange;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _colorUtils.rgbToHex;
  }
});
Object.defineProperty(exports, "runGpuFilterForPlot", {
  enumerable: true,
  get: function get() {
    return _plot.runGpuFilterForPlot;
  }
});
Object.defineProperty(exports, "scaleMapStyleByResolution", {
  enumerable: true,
  get: function get() {
    return _mapboxGlStyleEditor.scaleMapStyleByResolution;
  }
});
Object.defineProperty(exports, "selectRestCategoricalColorMapByIndex", {
  enumerable: true,
  get: function get() {
    return _dataScaleUtils.selectRestCategoricalColorMapByIndex;
  }
});
Object.defineProperty(exports, "setLayerBlending", {
  enumerable: true,
  get: function get() {
    return _glUtils.setLayerBlending;
  }
});
Object.defineProperty(exports, "setStyleSheetBaseHref", {
  enumerable: true,
  get: function get() {
    return _domUtils.setStyleSheetBaseHref;
  }
});
Object.defineProperty(exports, "snapToMarks", {
  enumerable: true,
  get: function get() {
    return _plot.snapToMarks;
  }
});
Object.defineProperty(exports, "sortCustomPaletteColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.sortCustomPaletteColor;
  }
});
Object.defineProperty(exports, "successNotification", {
  enumerable: true,
  get: function get() {
    return _notificationsUtils.successNotification;
  }
});
Object.defineProperty(exports, "transformRequest", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils2.transformRequest;
  }
});
Object.defineProperty(exports, "unobserveDimensions", {
  enumerable: true,
  get: function get() {
    return _observeDimensions.unobserveDimensions;
  }
});
Object.defineProperty(exports, "updateColorRangeByMatchingPalette", {
  enumerable: true,
  get: function get() {
    return _colorUtils.updateColorRangeByMatchingPalette;
  }
});
Object.defineProperty(exports, "updateColorRangeBySelectedPalette", {
  enumerable: true,
  get: function get() {
    return _colorUtils.updateColorRangeBySelectedPalette;
  }
});
Object.defineProperty(exports, "updateCustomColorRangeByColorUI", {
  enumerable: true,
  get: function get() {
    return _colorUtils.updateCustomColorRangeByColorUI;
  }
});
Object.defineProperty(exports, "updateCustomPaletteColor", {
  enumerable: true,
  get: function get() {
    return _colorUtils.updateCustomPaletteColor;
  }
});
Object.defineProperty(exports, "updateTimeFilterPlotType", {
  enumerable: true,
  get: function get() {
    return _plot.updateTimeFilterPlotType;
  }
});
Object.defineProperty(exports, "useDimensions", {
  enumerable: true,
  get: function get() {
    return _observeDimensions["default"];
  }
});
Object.defineProperty(exports, "validateEffectParameters", {
  enumerable: true,
  get: function get() {
    return _effectUtils.validateEffectParameters;
  }
});
Object.defineProperty(exports, "validateInputData", {
  enumerable: true,
  get: function get() {
    return _datasetUtils.validateInputData;
  }
});
Object.defineProperty(exports, "validateToken", {
  enumerable: true,
  get: function get() {
    return _mapboxUtils.validateToken;
  }
});
var _colorUtils = require("./color-utils");
var _notificationsUtils = require("./notifications-utils");
var _domUtils = require("./dom-utils");
var _domToImage = _interopRequireDefault(require("./dom-to-image"));
var _aggregation = require("./aggregation");
var _plot = require("./plot");
var _dataUtils = require("./data-utils");
Object.keys(_dataUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dataUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _dataUtils[key];
    }
  });
});
var _strings = require("./strings");
Object.keys(_strings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _strings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _strings[key];
    }
  });
});
var _time = require("./time");
var _positionUtils = require("./position-utils");
var _datasetUtils = require("./dataset-utils");
var _exportMapHtml = require("./export-map-html");
var _exportUtils = _interopRequireWildcard(require("./export-utils"));
var _format = require("./format");
var _glUtils = require("./gl-utils");
var _localeUtils = require("./locale-utils");
var _mapInfoUtils = require("./map-info-utils");
var _mapboxGlStyleEditor = require("./map-style-utils/mapbox-gl-style-editor");
var _mapboxUtils = require("./mapbox-utils");
var _observeDimensions = _interopRequireWildcard(require("./observe-dimensions"));
var _projectionUtils = require("./projection-utils");
Object.keys(_projectionUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _projectionUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _projectionUtils[key];
    }
  });
});
var _splitMapUtils = require("./split-map-utils");
Object.keys(_splitMapUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _splitMapUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _splitMapUtils[key];
    }
  });
});
var _utils = require("./utils");
Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _utils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _utils[key];
    }
  });
});
var _effectUtils = require("./effect-utils");
var _mapboxUtils2 = require("./map-style-utils/mapbox-utils");
var _mapUtils = require("./map-utils");
Object.keys(_mapUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mapUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _mapUtils[key];
    }
  });
});
var _arrowDataContainer = require("./arrow-data-container");
var _dataContainerUtils = require("./data-container-utils");
var _filterUtils = require("./filter-utils");
Object.keys(_filterUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _filterUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _filterUtils[key];
    }
  });
});
var _dataScaleUtils = require("./data-scale-utils");
var _dataRow = require("./data-row");
var _applicationConfig = require("./application-config");
var _browserUtils = require("./browser-utils");
var _quickInsertionSort = _interopRequireDefault(require("./quick-insertion-sort"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
