// @kepler.gl/layers/dist/trip-layer/trip-info-modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _reactMarkdown = _interopRequireDefault(require("react-markdown"));
var _reactIntl = require("react-intl");
var _localization = require("@kepler.gl/localization");
var _exampleTable = _interopRequireDefault(require("../example-table"));
var _templateObject, _templateObject2, _templateObject3; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var InfoModal = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 13px;\n  color: ", ";\n\n  pre {\n    padding: 12px;\n    background-color: #f8f8f9;\n  }\n"])), function (props) {
  return props.theme.titleColorLT;
});
var StyledTitle = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 20px;\n  letter-spacing: 1.25px;\n  margin: 18px 0 14px 0;\n  color: ", ";\n"])), function (props) {
  return props.theme.titleColorLT;
});
var StyledCode = _styledComponents["default"].code(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n"])), function (props) {
  return props.theme.titleColorLT;
});
var codeExampleGeojson = "\n".concat('```json', "\n{\n  \"type\": \"FeatureCollection\",\n  \"features\": [\n    {\n      \"type\": \"Feature\",\n      \"properties\": { \n        \"vendor\":  \"A\",\n        \"vol\":20\n      },\n      \"geometry\": {\n        \"type\": \"LineString\",\n        \"coordinates\": [\n          [-74.20986, 40.81773, 0, 1564184363],\n          [-74.20987, 40.81765, 0, 1564184396],\n          [-74.20998, 40.81746, 0, 1564184409]\n        ]\n      }\n    }\n  ]\n}\n", '```', "\n");
var exampleTableHeader = ['id', 'latitude', 'longitude', 'timestamp'];
var exampleTabbleRows = [['A', '40.81773', '-74.20986', '1564184363'], ['A', '40.81765', '-74.20987', '1564184396'], ['A', '40.81746', '-74.20998', '1564184409'], ['B', '40.64375', '-74.33242', '1565578213'], ['B', '40.64353', '-74.20987', '1565578217'], ['B', '40.64222', '-74.33001', '1565578243']];
var ExampleTable = function ExampleTable() {
  return /*#__PURE__*/_react["default"].createElement(_exampleTable["default"], {
    className: "trip-example-table"
  }, /*#__PURE__*/_react["default"].createElement("thead", null, /*#__PURE__*/_react["default"].createElement("tr", null, exampleTableHeader.map(function (v) {
    return /*#__PURE__*/_react["default"].createElement("th", {
      key: v
    }, v);
  }))), /*#__PURE__*/_react["default"].createElement("tbody", null, exampleTabbleRows.map(function (row, i) {
    return /*#__PURE__*/_react["default"].createElement("tr", {
      key: i
    }, row.map(function (v, j) {
      return /*#__PURE__*/_react["default"].createElement("td", {
        key: j
      }, /*#__PURE__*/_react["default"].createElement(StyledCode, null, v));
    }));
  })));
};
var TripInfoModalFactory = function TripInfoModalFactory(columnMode) {
  var TripInfoModal = function TripInfoModal() {
    var intl = (0, _reactIntl.useIntl)();
    return /*#__PURE__*/_react["default"].createElement(InfoModal, {
      className: "trip-info-modal"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "trip-info-modal__description"
    }, /*#__PURE__*/_react["default"].createElement(_reactMarkdown["default"], {
      children: intl.formatMessage({
        id: columnMode === 'geojson' ? 'modal.tripInfo.description1' : 'modal.tripInfo.descriptionTable1'
      })
    })), /*#__PURE__*/_react["default"].createElement("div", {
      className: "trip-info-modal__example"
    }, /*#__PURE__*/_react["default"].createElement(StyledTitle, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: columnMode === 'geojson' ? 'modal.tripInfo.example' : 'modal.tripInfo.exampleTable'
    })), columnMode === 'geojson' ? /*#__PURE__*/_react["default"].createElement(_reactMarkdown["default"], {
      children: codeExampleGeojson
    }) : /*#__PURE__*/_react["default"].createElement(ExampleTable, null)));
  };
  return TripInfoModal;
};
var _default = exports["default"] = TripInfoModalFactory;
