// @loaders.gl/arrow/dist/index.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
export { VECTOR_TYPES } from "./lib/types.js";
import { ArrowTableBatchAggregator } from "./schema/arrow-table-batch.js";
// Make the ArrowBatch type available
import { TableBatchBuilder } from '@loaders.gl/schema';
TableBatchBuilder.ArrowBatch = ArrowTableBatchAggregator;
export { ArrowWorkerLoader } from "./exports/arrow-loader.js";
export { ArrowLoader } from "./arrow-loader.js";
export { ArrowWriter } from "./arrow-writer.js";
// Geoarrow loader
export { GeoArrowWorkerLoader } from "./exports/geoarrow-loader.js";
export { GeoArrowLoader } from "./geoarrow-loader.js";
// Schema utils
export { convertArrowToSchema, convertSchemaToArrow, 
// DETAILED FUNCTIONS
serializeArrowSchema, deserializeArrowSchema, serializeArrowMetadata, deserializeArrowMetadata, serializeArrowField, deserializeArrowField, serializeArrowType, deserializeArrowType } from "./lib//tables/convert-arrow-schema.js";
// Table utils
export { convertArrowToTable } from "./lib/tables/convert-arrow-to-table.js";
export { convertTableToArrow } from "./lib/tables/convert-table-to-arrow.js";
export { getBinaryGeometryTemplate, getBinaryGeometriesFromArrow, getTriangleIndices, getMeanCentersFromBinaryGeometries } from "./lib/geoarrow/convert-geoarrow-to-binary-geometry.js";
export { updateBoundsFromGeoArrowSamples } from "./lib/geoarrow/get-arrow-bounds.js";
export { parseGeometryFromArrow } from "./lib/geoarrow/convert-geoarrow-to-geojson-geometry.js";
// EXPERIMENTAL WORKER
export { hardClone } from "./workers/hard-clone.js";
export { TriangulationWorker, triangulateOnWorker, parseGeoArrowOnWorker } from "./triangulate-on-worker.js";
