// @loaders.gl/arrow/dist/triangulate-on-worker.js
// loaders.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { processOnWorker } from '@loaders.gl/worker-utils';
// __VERSION__ is injected by babel-plugin-version-inline
// @ts-ignore TS2304: Cannot find name '__VERSION__'.
const VERSION = typeof "4.3.3" !== 'undefined' ? "4.3.3" : 'latest';
/**
 * Worker for tessellating geometries. Normally called through triangulateOnWorker
 */
export const TriangulationWorker = {
    id: 'triangulation',
    name: 'Triangulate',
    module: 'arrow',
    version: VERSION,
    options: {}
};
/**
 * Triangulate a set of polygons on worker, type safe API
 */
export function triangulateOnWorker(data, options = {}) {
    return processOnWorker(TriangulationWorker, { ...data, operation: 'triangulate' }, options);
}
/**
 * Parse GeoArrow geometry colum on worker, type safe API
 */
export function parseGeoArrowOnWorker(data, options = {}) {
    return processOnWorker(TriangulationWorker, { ...data, operation: 'parse-geoarrow' }, options);
}
