// @kepler.gl/components/dist/side-panel/layer-panel/color-palette-preset.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _d3Array = require("d3-array");
var _constants = require("@kepler.gl/constants");
var _templateObject, _templateObject2, _templateObject3; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var PALETTE_HEIGHT = '8px';
var ROWS = 22;
var StyledColorPalette = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: row;\n  justify-content: space-between;\n  padding: 12px;\n\n  &:hover {\n    cursor: pointer;\n  }\n"])));
var StyledColorColumn = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-grow: 1;\n  flex-direction: column;\n  justify-content: space-between;\n"])));
var StyledColorBlock = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  flex-grow: 1;\n  height: ", ";\n  border-width: 1px;\n  border-style: solid;\n"])), PALETTE_HEIGHT);
var PresetColorPalette = function PresetColorPalette(_ref) {
  var themes = _ref.themes,
    onSelectColor = _ref.onSelectColor,
    selectedColor = _ref.selectedColor;
  return /*#__PURE__*/_react["default"].createElement(StyledColorPalette, null, themes.map(function (theme) {
    return /*#__PURE__*/_react["default"].createElement(StyledColorColumn, {
      key: theme,
      className: "single-color-palette__column"
    }, (0, _d3Array.range)(1, ROWS + 1, 1).map(function (key) {
      return /*#__PURE__*/_react["default"].createElement(StyledColorBlock, {
        className: "single-color-palette__block",
        style: {
          backgroundColor: _constants.ColorsByTheme[theme][key],
          borderColor: selectedColor === _constants.ColorsByTheme[theme][key].toUpperCase() ? 'white' : _constants.ColorsByTheme[theme][key]
        },
        key: "".concat(theme, "_").concat(key),
        selected: selectedColor === _constants.ColorsByTheme[theme][key].toUpperCase(),
        onClick: function onClick(e) {
          return onSelectColor(_constants.ColorsByTheme[theme][key], e);
        }
      });
    }));
  }));
};
var _default = exports["default"] = PresetColorPalette;
