// @kepler.gl/deckgl-arrow-layers/dist/utils/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assignAccessor = assignAccessor;
exports.convertStructToFixedSizeList = convertStructToFixedSizeList;
exports.expandArrayToCoords = expandArrayToCoords;
exports.extractAccessorsFromProps = extractAccessorsFromProps;
exports.findGeometryColumnIndex = findGeometryColumnIndex;
exports.getGeometryVector = getGeometryVector;
exports.getListNestingLevels = getListNestingLevels;
exports.getMultiLineStringResolvedOffsets = getMultiLineStringResolvedOffsets;
exports.getMultiPolygonResolvedOffsets = getMultiPolygonResolvedOffsets;
exports.getPolygonResolvedOffsets = getPolygonResolvedOffsets;
exports.invertOffsets = invertOffsets;
exports.isColumnReference = isColumnReference;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _typed = require("@deck.gl/core/typed");
var arrow = _interopRequireWildcard(require("apache-arrow"));
var ga = _interopRequireWildcard(require("@geoarrow/geoarrow-js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

// deck.gl-community
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors

function findGeometryColumnIndex(schema, extensionName, geometryColumnName) {
  var index = schema.fields.findIndex(function (field) {
    return field.name === geometryColumnName || field.metadata.get('ARROW:extension:name') === extensionName;
  });
  return index !== -1 ? index : null;
}

/**
 * Returns `true` if the input is a reference to a column in the table
 */
function isColumnReference(input) {
  return typeof input === 'string';
}
function isDataInterleavedCoords(data) {
  // TODO: also check 2 or 3d? Float64?
  return data.type instanceof arrow.FixedSizeList;
}
function isDataSeparatedCoords(data) {
  // TODO: also check child names? Float64?
  return data.type instanceof arrow.Struct;
}

/**
 * Convert geoarrow Struct coordinates to FixedSizeList coords
 *
 * The GeoArrow spec allows for either separated or interleaved coords, but at
 * this time deck.gl only supports interleaved.
 */
// TODO: this hasn't been tested yet
function convertStructToFixedSizeList(coords) {
  if (isDataInterleavedCoords(coords)) {
    return coords;
  } else if (isDataSeparatedCoords(coords)) {
    // TODO: support 3d
    var interleavedCoords = new Float64Array(coords.length * 2);
    var _coords$children = (0, _slicedToArray2["default"])(coords.children, 2),
      xChild = _coords$children[0],
      yChild = _coords$children[1];
    for (var i = 0; i < coords.length; i++) {
      interleavedCoords[i * 2] = xChild.values[i];
      interleavedCoords[i * 2 + 1] = yChild.values[i];
    }
    var childDataType = new arrow.Float64();
    var dataType = new arrow.FixedSizeList(2, new arrow.Field('coords', childDataType));
    var interleavedCoordsData = arrow.makeData({
      type: childDataType,
      length: interleavedCoords.length
    });
    var data = arrow.makeData({
      type: dataType,
      length: coords.length,
      nullCount: coords.nullCount,
      nullBitmap: coords.nullBitmap,
      child: interleavedCoordsData
    });
    return data;
  }
  (0, _typed.assert)(false);
}
/**
 * A wrapper around a user-provided accessor function
 *
 * For layers like Scatterplot, Path, and Polygon, we automatically handle
 * "exploding" the table when multi-geometry input are provided. This means that
 * the upstream `index` value passed to the user will be the correct row index
 * _only_ for non-exploded data.
 *
 * With this function, we simplify the user usage by automatically converting
 * back from "exploded" index back to the original row index.
 */
function wrapAccessorFunction(objectInfo, userAccessorFunction, batchOffset) {
  var index = objectInfo.index,
    data = objectInfo.data;
  var newIndex = index + batchOffset;
  if (data.invertedGeomOffsets !== undefined) {
    newIndex = data.invertedGeomOffsets[index];
  }
  var newObjectData = {
    data: data.data,
    length: data.length,
    attributes: data.attributes
  };
  var newObjectInfo = {
    index: newIndex,
    data: newObjectData,
    target: objectInfo.target
  };
  return userAccessorFunction(newObjectInfo);
}

/**
 * Resolve accessor and assign to props object
 *
 * This is useful as a helper function because a scalar prop is set at the top
 * level while a vectorized prop is set inside data.attributes
 *
 */
function assignAccessor(args) {
  var props = args.props,
    propName = args.propName,
    propInput = args.propInput,
    chunkIdx = args.chunkIdx,
    geomCoordOffsets = args.geomCoordOffsets,
    _args$batchOffset = args.batchOffset,
    batchOffset = _args$batchOffset === void 0 ? 0 : _args$batchOffset;
  if (propInput === undefined) {
    return;
  }
  if (propInput instanceof arrow.Vector) {
    var columnData = propInput.data[chunkIdx];
    if (arrow.DataType.isFixedSizeList(columnData)) {
      (0, _typed.assert)(columnData.children.length === 1);
      var values = columnData.children[0].values;
      if (geomCoordOffsets) {
        values = expandArrayToCoords(values, columnData.type.listSize, geomCoordOffsets);
      }
      props.data.attributes[propName] = {
        value: values,
        size: columnData.type.listSize,
        // Set to `true` to signify that colors are already 0-255, and deck/luma
        // does not need to rescale
        // https://github.com/visgl/deck.gl/blob/401d624c0529faaa62125714c376b3ba3b8f379f/docs/api-reference/core/attribute-manager.md?plain=1#L66
        normalized: true
      };
    } else if (arrow.DataType.isFloat(columnData)) {
      var _values = columnData.values;
      if (geomCoordOffsets) {
        _values = expandArrayToCoords(_values, 1, geomCoordOffsets);
      }
      props.data.attributes[propName] = {
        value: _values,
        size: 1
      };
    }
  } else if (typeof propInput === 'function') {
    props[propName] = function (object, objectInfo) {
      // Special case that doesn't have the same parameters
      if (propName === 'getPolygonOffset') {
        return propInput(object, objectInfo);
      }
      return wrapAccessorFunction(objectInfo, propInput, batchOffset);
    };
  } else {
    props[propName] = propInput;
  }
}

/**
 * Expand an array from "one element per geometry" to "one element per coordinate"
 *
 * @param input: the input array to expand
 * @param size : the number of nested elements in the input array per geometry. So for example, for RGB data this would be 3, for RGBA this would be 4. For radius, this would be 1.
 * @param geomOffsets : an offsets array mapping from the geometry to the coordinate indexes. So in the case of a LineStringArray, this is retrieved directly from the GeoArrow storage. In the case of a PolygonArray, this comes from the resolved indexes that need to be given to the SolidPolygonLayer anyways.
 * @param numPositions : end position in geomOffsets, as geomOffsets can potentially contain preallocated zeroes in the end of the buffer.
 *
 * @return  {TypedArray} values expanded to be per-coordinate
 */
function expandArrayToCoords(input, size, geomOffsets, numPositions) {
  var lastIndex = numPositions || geomOffsets.length - 1;
  var numCoords = geomOffsets[lastIndex];
  // @ts-expect-error
  var outputArray = new input.constructor(numCoords * size);

  // geomIdx is an index into the geomOffsets array
  // geomIdx is also the geometry/table index
  for (var geomIdx = 0; geomIdx < lastIndex; geomIdx++) {
    // geomOffsets maps from the geometry index to the coord index
    // So here we get the range of coords that this geometry covers
    var lastCoordIdx = geomOffsets[geomIdx];
    var nextCoordIdx = geomOffsets[geomIdx + 1];

    // Iterate over this range of coord indices
    for (var coordIdx = lastCoordIdx; coordIdx < nextCoordIdx; coordIdx++) {
      // Iterate over size
      for (var i = 0; i < size; i++) {
        // Copy from the geometry index in `input` to the coord index in
        // `output`
        outputArray[coordIdx * size + i] = input[geomIdx * size + i];
      }
    }
  }
  return outputArray;
}

/**
 * Get a geometry vector with the specified extension type name from the table.
 */
function getGeometryVector(table, geoarrowTypeName) {
  var geometryColumnIdx = findGeometryColumnIndex(table.schema, geoarrowTypeName);
  if (geometryColumnIdx === null) {
    return null;
    // throw new Error(`No column found with extension type ${geoarrowTypeName}`);
  }
  return table.getChildAt(geometryColumnIdx);
}
function getListNestingLevels(data) {
  var nestingLevels = 0;
  if (arrow.DataType.isList(data.type)) {
    nestingLevels += 1;
    data = data.children[0];
  }
  return nestingLevels;
}
function getMultiLineStringResolvedOffsets(data) {
  var geomOffsets = data.valueOffsets;
  var lineStringData = ga.child.getMultiLineStringChild(data);
  var ringOffsets = lineStringData.valueOffsets;
  var resolvedRingOffsets = new Int32Array(geomOffsets.length);
  for (var i = 0; i < resolvedRingOffsets.length; ++i) {
    // Perform the lookup into the ringIndices array using the geomOffsets
    // array
    resolvedRingOffsets[i] = ringOffsets[geomOffsets[i]];
  }
  return resolvedRingOffsets;
}
function getPolygonResolvedOffsets(data) {
  var geomOffsets = data.valueOffsets;
  var ringData = ga.child.getPolygonChild(data);
  var ringOffsets = ringData.valueOffsets;
  var resolvedRingOffsets = new Int32Array(geomOffsets.length);
  for (var i = 0; i < resolvedRingOffsets.length; ++i) {
    // Perform the lookup into the ringIndices array using the geomOffsets
    // array
    resolvedRingOffsets[i] = ringOffsets[geomOffsets[i]];
  }
  return resolvedRingOffsets;
}
function getMultiPolygonResolvedOffsets(data) {
  var polygonData = ga.child.getMultiPolygonChild(data);
  var ringData = ga.child.getPolygonChild(polygonData);
  var geomOffsets = data.valueOffsets;
  var polygonOffsets = polygonData.valueOffsets;
  var ringOffsets = ringData.valueOffsets;
  var resolvedRingOffsets = new Int32Array(geomOffsets.length);
  for (var i = 0; i < resolvedRingOffsets.length; ++i) {
    resolvedRingOffsets[i] = ringOffsets[polygonOffsets[geomOffsets[i]]];
  }
  return resolvedRingOffsets;
}

/**
 * Invert offsets so that lookup can go in the opposite direction
 */
function invertOffsets(offsets) {
  var largestOffset = offsets[offsets.length - 1];
  var arrayConstructor = offsets.length < Math.pow(2, 8) ? Uint8Array : offsets.length < Math.pow(2, 16) ? Uint16Array : Uint32Array;
  var invertedOffsets = new arrayConstructor(largestOffset);
  for (var arrayIdx = 0; arrayIdx < offsets.length - 1; arrayIdx++) {
    var thisOffset = offsets[arrayIdx];
    var nextOffset = offsets[arrayIdx + 1];
    for (var offset = thisOffset; offset < nextOffset; offset++) {
      invertedOffsets[offset] = arrayIdx;
    }
  }
  return invertedOffsets;
}

// TODO: better typing
function extractAccessorsFromProps(props, excludeKeys) {
  var accessors = {};
  var otherProps = {};
  for (var _i = 0, _Object$entries = Object.entries(props); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = (0, _slicedToArray2["default"])(_Object$entries[_i], 2),
      key = _Object$entries$_i[0],
      value = _Object$entries$_i[1];
    if (excludeKeys.includes(key)) {
      continue;
    }
    if (key.startsWith('get')) {
      accessors[key] = value;
    } else {
      otherProps[key] = value;
    }
  }
  return [accessors, otherProps];
}
