// @kepler.gl/components/dist/modals/tilesets-modals/load-data-footer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireDefault(require("react"));
var _reactIntl = require("react-intl");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _common = require("../../common");
var _loadingSpinner = _interopRequireDefault(require("../../common/loading-spinner"));
var _templateObject, _templateObject2, _templateObject3; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var AddDataButton = (0, _styledComponents["default"])(_common.Button)(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  position: relative;\n  display: inline-flex;\n  align-items: center;\n  opacity: ", ";\n  height: 40px;\n  padding: 9px 32px;\n  width: ", ";\n\n  svg {\n    margin-right: ", ";\n  }\n"])), function (props) {
  return props.disabled ? 0.6 : 1;
}, function (props) {
  return props.width;
}, function (props) {
  return props.isLoading ? '0' : '6px';
});
var LoadDataFooterContainer = _styledComponents["default"].div.attrs({
  className: 'load-data-footer'
})(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  /* position: absolute; */\n  width: 100%;\n  bottom: 0;\n\n  display: flex;\n  justify-content: space-between;\n  padding: 21px 21px 0px 72px;\n  margin: 24px -72px 0;\n  align-items: center;\n"])));
var ErrorContainer = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  color: red;\n  padding-left: 15px;\n  display: inline-block;\n"])));
var LoadDataFooter = function LoadDataFooter(_ref) {
  var disabled = _ref.disabled,
    intl = _ref.intl,
    isLoading = _ref.isLoading,
    onConfirm = _ref.onConfirm,
    confirmText = _ref.confirmText,
    _ref$prependText = _ref.prependText,
    prependText = _ref$prependText === void 0 ? '' : _ref$prependText,
    _ref$errorText = _ref.errorText,
    errorText = _ref$errorText === void 0 ? '' : _ref$errorText;
  return /*#__PURE__*/_react["default"].createElement(LoadDataFooterContainer, null, /*#__PURE__*/_react["default"].createElement("div", null, prependText, /*#__PURE__*/_react["default"].createElement(AddDataButton, {
    disabled: disabled,
    isLoading: isLoading,
    onClick: onConfirm,
    width: "130px",
    cta: true
  }, isLoading && /*#__PURE__*/_react["default"].createElement(_loadingSpinner["default"], {
    size: 12
  }), isLoading ? null : intl.formatMessage({
    id: confirmText
  })), errorText && /*#__PURE__*/_react["default"].createElement(ErrorContainer, null, errorText)));
};
var _default = exports["default"] = (0, _reactIntl.injectIntl)(LoadDataFooter);
