// @kepler.gl/constants/dist/dataset.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RemoteTileFormat = exports.RasterTileType = exports.REMOTE_TILE = exports.PMTilesType = exports.DatasetType = void 0;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var DatasetType = exports.DatasetType = /*#__PURE__*/function (DatasetType) {
  DatasetType["LOCAL"] = "local";
  DatasetType["VECTOR_TILE"] = "vector-tile";
  DatasetType["RASTER_TILE"] = "raster-tile";
  DatasetType["WMS_TILE"] = "wms-tile";
  return DatasetType;
}({});
var RemoteTileFormat = exports.RemoteTileFormat = /*#__PURE__*/function (RemoteTileFormat) {
  RemoteTileFormat["MVT"] = "mvt";
  RemoteTileFormat["PMTILES"] = "pmtiles";
  RemoteTileFormat["WMS"] = "wms";
  return RemoteTileFormat;
}({});
var PMTilesType = exports.PMTilesType = /*#__PURE__*/function (PMTilesType) {
  PMTilesType["RASTER"] = "raster";
  PMTilesType["MVT"] = "mvt";
  return PMTilesType;
}({});
var REMOTE_TILE = exports.REMOTE_TILE = 'remote';

/**
 * Raster tileset metadata in STAC Item format. STAC version must be >= 1.0.0,
 * and the EO and Raster STAC extensions are required. This metadata shape can
 * be passed to the map to synchronously add a raster tileset.
 * @see https://github.com/radiantearth/stac-spec/blob/master/item-spec/item-spec.md
 */

/**
 * Raster tileset metadata with a remote metadata URL. This metadata can
 * be passed to the map to asynchronously load a raster tileset.
 */
var RasterTileType = exports.RasterTileType = /*#__PURE__*/function (RasterTileType) {
  RasterTileType["STAC"] = "stac";
  RasterTileType["PMTILES"] = "pmtiles";
  return RasterTileType;
}({});
/**
 * Raster tileset metadata.
 */
